{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc01d59ced195153",
   "metadata": {},
   "source": [
    "# Biolink Model Subsetting\n",
    "This notebook demonstrates how to subset the Biolink Model using the Biolink Model schema and a transformation specification in two different ways: first using a transformation specification defined manually in YAML and loaded from the filesystem, and second, using a subset of classes retrieved using SchemaView dynamically."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b3677392f95575",
   "metadata": {},
   "source": [
    "First, we import several LinkML helper packages including SchemaView: https://linkml.io/linkml/developers/schemaview.html \n",
    "SchemaView is a LinkML schema introspection tool that provides helpful functions to retrieve model elements by name.  It also supports navigating ancestors and descendants of any model element, including classes, slots, types, and permissible values in enumerations.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e3f86775ef7a9dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from pytest import fixture\n",
    "\n",
    "from linkml_runtime.dumpers import yaml_dumper\n",
    "from linkml_map.datamodel.transformer_model import TransformationSpecification, ClassDerivation, SlotDerivation, CopyDirective\n",
    "from linkml_map.inference.schema_mapper import SchemaMapper\n",
    "from linkml_map.session import Session\n",
    "from linkml_runtime.utils.schemaview import SchemaView\n",
    "from linkml_map.utils.loaders import load_specification\n",
    "from linkml_runtime.utils.formatutils import camelcase, underscore\n",
    "from pprint import pprint\n",
    "\n",
    "REPO_ROOT = Path.cwd().parent.parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d3b8f7bc023dae",
   "metadata": {},
   "source": [
    "SchemaView can accept a variety of imports to initialize, including a Path or string representation of a Path or (as we do in this case), a URL of a raw LinkML schema. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8ca1d183bd948863",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_url = \"https://raw.githubusercontent.com/biolink/biolink-model/master/biolink-model.yaml\"\n",
    "sv = SchemaView(schema_url)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91dc68f7d162c6f",
   "metadata": {},
   "source": [
    "### Creating a Transformation Specification Manually in YAML format.\n",
    "In our first example, we develop a \"Transform Specification\" for Biolink Model using Class and Slot Derivations as defined by the linkml-map transformation language.  More about that here: https://linkml.io/linkml-map/#TransformationSpecification/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b9f39d577790a0dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NamedThing\n",
      "Gene\n",
      "Disease\n",
      "PhenotypicFeature\n",
      "Association\n",
      "GeneToPhenotypicFeatureAssociation\n",
      "\n",
      "id\n",
      "name\n",
      "category\n",
      "symbol\n",
      "subject\n",
      "predicate\n",
      "object\n"
     ]
    }
   ],
   "source": [
    "transform_file = REPO_ROOT / \"tests/input/examples/biolink/transform/biolink-example-profile.transform.yaml\"\n",
    "# Initialize Session and SchemaBuilder\n",
    "session = Session()\n",
    "\n",
    "# Set the source schema in the session\n",
    "session.set_source_schema(sv)\n",
    "\n",
    "tr_spec = load_specification(transform_file)\n",
    "mapper = SchemaMapper()\n",
    "mapper.source_schemaview = sv\n",
    "\n",
    "target_schema_obj = mapper.derive_schema(specification=tr_spec,\n",
    "                                         target_schema_id=\"biolink-profile\",\n",
    "                                        target_schema_name=\"BiolinkProfile\")\n",
    "\n",
    "\n",
    "yaml_dumper.dump(target_schema_obj, str(\"biolink-profile.yaml\"))\n",
    "\n",
    "transformed_sv = SchemaView(\"biolink-profile.yaml\")\n",
    "\n",
    "for class_name in transformed_sv.all_classes():\n",
    "    print(class_name)\n",
    "print()\n",
    "for slot_name in transformed_sv.all_slots():\n",
    "    print(slot_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8ac53bfef2f51db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception: range: label type\n"
     ]
    }
   ],
   "source": [
    "!gen-pydantic biolink-profile.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58dcaf83d20f4dea",
   "metadata": {},
   "source": [
    "Note:\n",
    "* Still in development is the tracing of provenance between a source schema and a destination schema.  Right now there is no provenance.  \n",
    "* If a class, slot, enum, or type is not included at all in the Derivation, it will not be pushed forward to the destination schema.  In development is an option to pull all non-specified components of the source model into the destination model.\n",
    "* You can do transformations on a derivation as well, see: https://linkml.io/linkml-map/#examples/Tutorial/#using-expressions\n",
    "* You can transform data as well as schemas, but this is currently \"beta\" level development.\n",
    "* Custom types are not pulled forward; this is a result of an error at the moment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d85f450751e59",
   "metadata": {},
   "source": [
    "### Creating a TransformationSpecification using SchemaView and an existing model.\n",
    "\n",
    "In our second example, we use the Biolink Model directly to derive classes and slots programmatically according to a simple list of \"subset classes\" that we want to extract from the main model in order to produce a subset model according to our specification."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a7c5a345c5f87d",
   "metadata": {},
   "source": [
    "First, we write a method to extract the classes and slots from Biolink Model using SchemaView"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "87000f75753c5faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_biolink_class_derivations(sv, subset_classes) -> dict:\n",
    "    \"\"\"\n",
    "    Function to get Biolink class definitions\n",
    "\n",
    "    :param sv: SchemaView object\n",
    "    :param subset_classes: List of classes to subset\n",
    "    :return: Dictionary of class derivations incl slot derivations\n",
    "    \"\"\"\n",
    "    # Example implementation to fetch class definitions\n",
    "    # This should be replaced with the actual implementation\n",
    "    class_derivations ={}\n",
    "    for class_name in subset_classes:\n",
    "        class_derivation = ClassDerivation(populated_from=class_name,\n",
    "                                           name=camelcase(class_name))\n",
    "        induced_slots = sv.class_induced_slots(class_name)\n",
    "        for slot in induced_slots:\n",
    "            slot_derivation = SlotDerivation(populated_from=slot.name, name=underscore(slot.name))\n",
    "            class_derivation.slot_derivations[underscore(slot.name)] = slot_derivation\n",
    "        class_derivations[camelcase(class_name)] = class_derivation\n",
    "    return class_derivations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1193a55025345761",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'is_a'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[40], line 27\u001b[0m\n\u001b[1;32m     24\u001b[0m mapper \u001b[38;5;241m=\u001b[39m SchemaMapper()\n\u001b[1;32m     25\u001b[0m mapper\u001b[38;5;241m.\u001b[39msource_schemaview \u001b[38;5;241m=\u001b[39m sv\n\u001b[0;32m---> 27\u001b[0m target_schema_obj \u001b[38;5;241m=\u001b[39m \u001b[43mmapper\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mderive_schema\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m    \u001b[49m\u001b[43mspecification\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_schema_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbiolink-subset\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_schema_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mBiolinkSubset\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     29\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m     31\u001b[0m yaml_dumper\u001b[38;5;241m.\u001b[39mdump(target_schema_obj, \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbiolink-subset.yaml\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m     33\u001b[0m transformed_sv \u001b[38;5;241m=\u001b[39m SchemaView(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbiolink-subset.yaml\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Documents/src/linkml-map/src/linkml_map/inference/schema_mapper.py:181\u001b[0m, in \u001b[0;36mSchemaMapper.derive_schema\u001b[0;34m(self, specification, target_schema_id, target_schema_name, suffix)\u001b[0m\n\u001b[1;32m    179\u001b[0m target_schema\u001b[38;5;241m.\u001b[39mdefault_range \u001b[38;5;241m=\u001b[39m source_schema\u001b[38;5;241m.\u001b[39mdefault_range\n\u001b[1;32m    180\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m cd \u001b[38;5;129;01min\u001b[39;00m target_schema\u001b[38;5;241m.\u001b[39mclasses\u001b[38;5;241m.\u001b[39mvalues():\n\u001b[0;32m--> 181\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_rewire_class\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcd\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    182\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m (cn, sn), info \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mslot_info\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m    183\u001b[0m     cd \u001b[38;5;241m=\u001b[39m target_schema\u001b[38;5;241m.\u001b[39mclasses[cn]\n",
      "File \u001b[0;32m~/Documents/src/linkml-map/src/linkml_map/inference/schema_mapper.py:323\u001b[0m, in \u001b[0;36mSchemaMapper._rewire_class\u001b[0;34m(self, class_definition)\u001b[0m\n\u001b[1;32m    321\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_rewire_class\u001b[39m(\u001b[38;5;28mself\u001b[39m, class_definition: ClassDefinition):\n\u001b[1;32m    322\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m class_definition\u001b[38;5;241m.\u001b[39mis_a:\n\u001b[0;32m--> 323\u001b[0m         class_definition\u001b[38;5;241m.\u001b[39mis_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_rewire_parent\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclass_definition\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mclass_definition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_a\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    324\u001b[0m     mixins \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_rewire_parent(class_definition, m) \u001b[38;5;28;01mfor\u001b[39;00m m \u001b[38;5;129;01min\u001b[39;00m class_definition\u001b[38;5;241m.\u001b[39mmixins]\n\u001b[1;32m    325\u001b[0m     class_definition\u001b[38;5;241m.\u001b[39mmixins \u001b[38;5;241m=\u001b[39m [m \u001b[38;5;28;01mfor\u001b[39;00m m \u001b[38;5;129;01min\u001b[39;00m mixins \u001b[38;5;28;01mif\u001b[39;00m m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m]\n",
      "File \u001b[0;32m~/Documents/src/linkml-map/src/linkml_map/inference/schema_mapper.py:339\u001b[0m, in \u001b[0;36mSchemaMapper._rewire_parent\u001b[0;34m(self, class_definition, parent)\u001b[0m\n\u001b[1;32m    337\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m new_parents[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    338\u001b[0m parent_cls \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msource_schemaview\u001b[38;5;241m.\u001b[39mget_class(parent)\n\u001b[0;32m--> 339\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mparent_cls\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_a\u001b[49m:\n\u001b[1;32m    340\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_rewire_parent(class_definition, parent_cls\u001b[38;5;241m.\u001b[39mis_a)\n\u001b[1;32m    341\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'is_a'"
     ]
    }
   ],
   "source": [
    "session = Session()\n",
    "\n",
    "# Set the source schema in the session\n",
    "session.set_source_schema(sv)\n",
    "\n",
    "subset_classes = [\n",
    "        \"gene\",\n",
    "        \"disease\",\n",
    "        \"case to phenotypic feature association\",\n",
    "        \"gene to disease association\",\n",
    "        \"gene to phenotypic feature association\",\n",
    "        \"case\",\n",
    "        \"phenotypic feature\",\n",
    "    ]\n",
    "\n",
    "class_derivations = get_biolink_class_derivations(sv, subset_classes)\n",
    "copy_type_directives = {\n",
    "    type_name: CopyDirective(element_name=type_name, copy_all=True)\n",
    "    for type_name, type_def in sv.all_types().items()\n",
    "}\n",
    "\n",
    "ts = TransformationSpecification(class_derivations=class_derivations, copy_directives=copy_type_directives)\n",
    "\n",
    "mapper = SchemaMapper()\n",
    "mapper.source_schemaview = sv\n",
    "\n",
    "target_schema_obj = mapper.derive_schema(\n",
    "    specification=ts, target_schema_id=\"biolink-subset\", target_schema_name=\"BiolinkSubset\"\n",
    ")\n",
    "\n",
    "yaml_dumper.dump(target_schema_obj, str(\"biolink-subset.yaml\"))\n",
    "\n",
    "transformed_sv = SchemaView(\"biolink-subset.yaml\")\n",
    "\n",
    "for class_name in transformed_sv.all_classes():\n",
    "    print(\"class derived: \", class_name)\n",
    "for slot_name in transformed_sv.all_slots():\n",
    "    print(\"slot derived: \", slot_name)\n",
    "for type_name in transformed_sv.all_types():\n",
    "    print(\"type copied: \", type_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f560d74e30c83f1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name: BiolinkSubset\n",
      "id: biolink-subset\n",
      "imports:\n",
      "- linkml:types\n",
      "prefixes:\n",
      "  AGRKB:\n",
      "    prefix_prefix: AGRKB\n",
      "    prefix_reference: https://www.alliancegenome.org/\n",
      "  apollo:\n",
      "    prefix_prefix: apollo\n",
      "    prefix_reference: https://github.com/GMOD/Apollo\n",
      "  AspGD:\n",
      "    prefix_prefix: AspGD\n",
      "    prefix_reference: http://www.aspergillusgenome.org/cgi-bin/locus.pl?dbid=\n",
      "  biolink:\n",
      "    prefix_prefix: biolink\n",
      "    prefix_reference: https://w3id.org/biolink/vocab/\n",
      "  bioschemas:\n",
      "    prefix_prefix: bioschemas\n",
      "    prefix_reference: https://bioschemas.org/\n",
      "  linkml:\n",
      "    prefix_prefix: linkml\n",
      "    prefix_reference: https://w3id.org/linkml/\n",
      "  CAID:\n",
      "    prefix_prefix: CAID\n",
      "    prefix_reference: http://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_caid?caid=\n",
      "  CHADO:\n",
      "    prefix_prefix: CHADO\n",
      "    prefix_reference: http://gmod.org/wiki/Chado/\n",
      "  ChemBank:\n",
      "    prefix_prefix: ChemBank\n",
      "    prefix_reference: http://chembank.broadinstitute.org/chemistry/viewMolecule.htm?cbid=\n",
      "  CHEMBL.MECHANISM:\n",
      "    prefix_prefix: CHEMBL.MECHANISM\n",
      "    prefix_reference: https://www.ebi.ac.uk/chembl/mechanism/inspect/\n",
      "  CID:\n",
      "    prefix_prefix: CID\n",
      "    prefix_reference: http://pubchem.ncbi.nlm.nih.gov/compound/\n",
      "  CLINVAR:\n",
      "    prefix_prefix: CLINVAR\n",
      "    prefix_reference: http://identifiers.org/clinvar\n",
      "  COAR_RESOURCE:\n",
      "    prefix_prefix: COAR_RESOURCE\n",
      "    prefix_reference: http://purl.org/coar/resource_type/\n",
      "  COG:\n",
      "    prefix_prefix: COG\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/research/cog-project/\n",
      "  ComplexPortal:\n",
      "    prefix_prefix: ComplexPortal\n",
      "    prefix_reference: https://www.ebi.ac.uk/complexportal/complex/\n",
      "  CPT:\n",
      "    prefix_prefix: CPT\n",
      "    prefix_reference: https://www.ama-assn.org/practice-management/cpt/\n",
      "  CTD.CHEMICAL:\n",
      "    prefix_prefix: CTD.CHEMICAL\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=chem&acc=\n",
      "  CTD.DISEASE:\n",
      "    prefix_prefix: CTD.DISEASE\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=disease&db=MESH&acc=\n",
      "  CTD.GENE:\n",
      "    prefix_prefix: CTD.GENE\n",
      "    prefix_reference: http://ctdbase.org/detail.go?type=gene&acc=\n",
      "  CTD:\n",
      "    prefix_prefix: CTD\n",
      "    prefix_reference: http://ctdbase.org/\n",
      "  DGIdb:\n",
      "    prefix_prefix: DGIdb\n",
      "    prefix_reference: https://www.dgidb.org/interaction_types\n",
      "  dcat:\n",
      "    prefix_prefix: dcat\n",
      "    prefix_reference: http://www.w3.org/ns/dcat#\n",
      "  dct:\n",
      "    prefix_prefix: dct\n",
      "    prefix_reference: http://purl.org/dc/terms/\n",
      "  dcid:\n",
      "    prefix_prefix: dcid\n",
      "    prefix_reference: https://datacommons.org/browser/\n",
      "  doi:\n",
      "    prefix_prefix: doi\n",
      "    prefix_reference: https://doi.org/\n",
      "  DOID-PROPERTY:\n",
      "    prefix_prefix: DOID-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/doid#\n",
      "  DrugCentral:\n",
      "    prefix_prefix: DrugCentral\n",
      "    prefix_reference: http://drugcentral.org/drugcard/\n",
      "  ECTO:\n",
      "    prefix_prefix: ECTO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ECTO_\n",
      "  EDAM-DATA:\n",
      "    prefix_prefix: EDAM-DATA\n",
      "    prefix_reference: http://edamontology.org/data_\n",
      "  EDAM-FORMAT:\n",
      "    prefix_prefix: EDAM-FORMAT\n",
      "    prefix_reference: http://edamontology.org/format_\n",
      "  EDAM-OPERATION:\n",
      "    prefix_prefix: EDAM-OPERATION\n",
      "    prefix_reference: http://edamontology.org/operation_\n",
      "  EDAM-TOPIC:\n",
      "    prefix_prefix: EDAM-TOPIC\n",
      "    prefix_reference: http://edamontology.org/topic_\n",
      "  EFO:\n",
      "    prefix_prefix: EFO\n",
      "    prefix_reference: http://www.ebi.ac.uk/efo/EFO_\n",
      "  ExO:\n",
      "    prefix_prefix: ExO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ExO_\n",
      "  fabio:\n",
      "    prefix_prefix: fabio\n",
      "    prefix_reference: http://purl.org/spar/fabio/\n",
      "  foaf:\n",
      "    prefix_prefix: foaf\n",
      "    prefix_reference: http://xmlns.com/foaf/0.1/\n",
      "  foodb.compound:\n",
      "    prefix_prefix: foodb.compound\n",
      "    prefix_reference: http://foodb.ca/compounds/\n",
      "  foodb.food:\n",
      "    prefix_prefix: foodb.food\n",
      "    prefix_reference: http://foodb.ca/foods/\n",
      "  FYECO:\n",
      "    prefix_prefix: FYECO\n",
      "    prefix_reference: https://www.pombase.org/term/\n",
      "  FYPO:\n",
      "    prefix_prefix: FYPO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/FYPO_\n",
      "  gff3:\n",
      "    prefix_prefix: gff3\n",
      "    prefix_reference: https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md#\n",
      "  GOREL:\n",
      "    prefix_prefix: GOREL\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/GOREL_\n",
      "  GOP:\n",
      "    prefix_prefix: GOP\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/go#\n",
      "  gpi:\n",
      "    prefix_prefix: gpi\n",
      "    prefix_reference: https://github.com/geneontology/go-annotation/blob/master/specs/gpad-gpi-2-0.md#\n",
      "  GSID:\n",
      "    prefix_prefix: GSID\n",
      "    prefix_reference: https://scholar.google.com/citations?user=\n",
      "  GTEx:\n",
      "    prefix_prefix: GTEx\n",
      "    prefix_reference: https://www.gtexportal.org/home/gene/\n",
      "  GTOPDB:\n",
      "    prefix_prefix: GTOPDB\n",
      "    prefix_reference: https://www.guidetopharmacology.org/GRAC/LigandDisplayForward?ligandId=\n",
      "  gtpo:\n",
      "    prefix_prefix: gtpo\n",
      "    prefix_reference: https://rdf.guidetopharmacology.org/ns/gtpo#\n",
      "  HANCESTRO:\n",
      "    prefix_prefix: HANCESTRO\n",
      "    prefix_reference: http://www.ebi.ac.uk/ancestro/ancestro_\n",
      "  HCPCS:\n",
      "    prefix_prefix: HCPCS\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/HCPCS/\n",
      "  HsapDv:\n",
      "    prefix_prefix: HsapDv\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/HsapDv_\n",
      "  ICD10:\n",
      "    prefix_prefix: ICD10\n",
      "    prefix_reference: https://icd.codes/icd9cm/\n",
      "  ICD9:\n",
      "    prefix_prefix: ICD9\n",
      "    prefix_reference: http://translator.ncats.nih.gov/ICD9_\n",
      "  interpro:\n",
      "    prefix_prefix: interpro\n",
      "    prefix_reference: https://www.ebi.ac.uk/interpro/entry/\n",
      "  INO:\n",
      "    prefix_prefix: INO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/INO_\n",
      "  isbn:\n",
      "    prefix_prefix: isbn\n",
      "    prefix_reference: https://www.isbn-international.org/identifier/\n",
      "  isni:\n",
      "    prefix_prefix: isni\n",
      "    prefix_reference: https://isni.org/isni/\n",
      "  issn:\n",
      "    prefix_prefix: issn\n",
      "    prefix_reference: https://portal.issn.org/resource/ISSN/\n",
      "  ncats.drug:\n",
      "    prefix_prefix: ncats.drug\n",
      "    prefix_reference: https://drugs.ncats.io/drug/\n",
      "  KEGG.BRITE:\n",
      "    prefix_prefix: KEGG.BRITE\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.brite:'\n",
      "  KEGG:\n",
      "    prefix_prefix: KEGG\n",
      "    prefix_reference: http://www.kegg.jp/entry/\n",
      "  KEGG.GENES:\n",
      "    prefix_prefix: KEGG.GENES\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.genes:bsu:'\n",
      "  KEGG.PATHWAY:\n",
      "    prefix_prefix: KEGG.PATHWAY\n",
      "    prefix_reference: 'https://bioregistry.io/kegg.pathway:'\n",
      "  KEGG.RCLASS:\n",
      "    prefix_prefix: KEGG.RCLASS\n",
      "    prefix_reference: 'https://www.genome.jp/dbget-bin/www_bget?rc:'\n",
      "  LOINC:\n",
      "    prefix_prefix: LOINC\n",
      "    prefix_reference: http://loinc.org/rdf/\n",
      "  MAXO:\n",
      "    prefix_prefix: MAXO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MAXO_\n",
      "  medgen:\n",
      "    prefix_prefix: medgen\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/medgen/\n",
      "  metacyc.reaction:\n",
      "    prefix_prefix: metacyc.reaction\n",
      "    prefix_reference: 'http://identifiers.org/metacyc.reaction:'\n",
      "  METANETX.REACTION:\n",
      "    prefix_prefix: METANETX.REACTION\n",
      "    prefix_reference: https://www.metanetx.org/equa_info/\n",
      "  MESH:\n",
      "    prefix_prefix: MESH\n",
      "    prefix_reference: http://id.nlm.nih.gov/mesh/\n",
      "  MI:\n",
      "    prefix_prefix: MI\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MI_\n",
      "  mirbase:\n",
      "    prefix_prefix: mirbase\n",
      "    prefix_reference: http://identifiers.org/mirbase\n",
      "  mmmp.biomaps:\n",
      "    prefix_prefix: mmmp.biomaps\n",
      "    prefix_reference: 'https://bioregistry.io/mmmp.biomaps:'\n",
      "  MmusDv:\n",
      "    prefix_prefix: MmusDv\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/MMUSDV_\n",
      "  MSigDB:\n",
      "    prefix_prefix: MSigDB\n",
      "    prefix_reference: https://www.gsea-msigdb.org/gsea/msigdb/\n",
      "  NBO-PROPERTY:\n",
      "    prefix_prefix: NBO-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/nbo#\n",
      "  ncats.bioplanet:\n",
      "    prefix_prefix: ncats.bioplanet\n",
      "    prefix_reference: https://tripod.nih.gov/bioplanet/detail.jsp?pid=\n",
      "  NCBIGene:\n",
      "    prefix_prefix: NCBIGene\n",
      "    prefix_reference: http://identifiers.org/ncbigene/\n",
      "  NCIT-OBO:\n",
      "    prefix_prefix: NCIT-OBO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/ncit#\n",
      "  NDDF:\n",
      "    prefix_prefix: NDDF\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/NDDF/\n",
      "  NLMID:\n",
      "    prefix_prefix: NLMID\n",
      "    prefix_reference: https://www.ncbi.nlm.nih.gov/nlmcatalog/?term=\n",
      "  OBAN:\n",
      "    prefix_prefix: OBAN\n",
      "    prefix_reference: http://purl.org/oban/\n",
      "  OMIM.PS:\n",
      "    prefix_prefix: OMIM.PS\n",
      "    prefix_reference: https://www.omim.org/phenotypicSeries/\n",
      "  ORCID:\n",
      "    prefix_prefix: ORCID\n",
      "    prefix_reference: https://orcid.org/\n",
      "  orphanet:\n",
      "    prefix_prefix: orphanet\n",
      "    prefix_reference: http://www.orpha.net/ORDO/Orphanet_\n",
      "  os:\n",
      "    prefix_prefix: os\n",
      "    prefix_reference: https://github.com/cmungall/owlstar/blob/master/owlstar.ttl\n",
      "  PANTHER.FAMILY:\n",
      "    prefix_prefix: PANTHER.FAMILY\n",
      "    prefix_reference: http://www.pantherdb.org/panther/family.do?clsAccession=\n",
      "  PathWhiz:\n",
      "    prefix_prefix: PathWhiz\n",
      "    prefix_reference: http://smpdb.ca/pathways/#\n",
      "  pav:\n",
      "    prefix_prefix: pav\n",
      "    prefix_reference: http://purl.org/pav/\n",
      "  PHARMGKB.DRUG:\n",
      "    prefix_prefix: PHARMGKB.DRUG\n",
      "    prefix_reference: https://www.pharmgkb.org/chemical/\n",
      "  PHARMGKB.DISEASE:\n",
      "    prefix_prefix: PHARMGKB.DISEASE\n",
      "    prefix_reference: https://www.pharmgkb.org/disease/\n",
      "  PHARMGKB.GENE:\n",
      "    prefix_prefix: PHARMGKB.GENE\n",
      "    prefix_reference: https://www.pharmgkb.org/gene/\n",
      "  PHARMGKB.PATHWAYS:\n",
      "    prefix_prefix: PHARMGKB.PATHWAYS\n",
      "    prefix_reference: https://www.pharmgkb.org/pathway/\n",
      "  PHARMGKB.VARIANT:\n",
      "    prefix_prefix: PHARMGKB.VARIANT\n",
      "    prefix_reference: https://www.pharmgkb.org/variant/\n",
      "  PHAROS:\n",
      "    prefix_prefix: PHAROS\n",
      "    prefix_reference: http://pharos.nih.gov\n",
      "  PomBase:\n",
      "    prefix_prefix: PomBase\n",
      "    prefix_reference: https://www.pombase.org/gene/\n",
      "  prov:\n",
      "    prefix_prefix: prov\n",
      "    prefix_reference: http://www.w3.org/ns/prov#\n",
      "  qud:\n",
      "    prefix_prefix: qud\n",
      "    prefix_reference: http://qudt.org/1.1/schema/qudt#\n",
      "  REPODB:\n",
      "    prefix_prefix: REPODB\n",
      "    prefix_reference: http://apps.chiragjpgroup.org/repoDB/\n",
      "  ResearchID:\n",
      "    prefix_prefix: ResearchID\n",
      "    prefix_reference: https://publons.com/researcher/\n",
      "  RO:\n",
      "    prefix_prefix: RO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/RO_\n",
      "  RXNORM:\n",
      "    prefix_prefix: RXNORM\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/RXNORM/\n",
      "  RXCUI:\n",
      "    prefix_prefix: RXCUI\n",
      "    prefix_reference: https://mor.nlm.nih.gov/RxNav/search?searchBy=RXCUI&searchTerm=\n",
      "  schema:\n",
      "    prefix_prefix: schema\n",
      "    prefix_reference: http://schema.org/\n",
      "  ScopusID:\n",
      "    prefix_prefix: ScopusID\n",
      "    prefix_reference: https://www.scopus.com/authid/detail.uri?authorId=\n",
      "  SEED.REACTION:\n",
      "    prefix_prefix: SEED.REACTION\n",
      "    prefix_reference: https://modelseed.org/biochem/reactions/\n",
      "  SEMMEDDB:\n",
      "    prefix_prefix: SEMMEDDB\n",
      "    prefix_reference: https://skr3.nlm.nih.gov/SemMedDB\n",
      "  SIO:\n",
      "    prefix_prefix: SIO\n",
      "    prefix_reference: http://semanticscience.org/resource/SIO_\n",
      "  SNOMEDCT:\n",
      "    prefix_prefix: SNOMEDCT\n",
      "    prefix_reference: http://snomed.info/id/\n",
      "  SPDI:\n",
      "    prefix_prefix: SPDI\n",
      "    prefix_reference: https://api.ncbi.nlm.nih.gov/variation/v0/spdi/\n",
      "  UBERGRAPH:\n",
      "    prefix_prefix: UBERGRAPH\n",
      "    prefix_reference: http://translator.renci.org/ubergraph-axioms.ofn#\n",
      "  UBERON_CORE:\n",
      "    prefix_prefix: UBERON_CORE\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/uberon/core#\n",
      "  UBERON_NONAMESPACE:\n",
      "    prefix_prefix: UBERON_NONAMESPACE\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/core#\n",
      "  STY:\n",
      "    prefix_prefix: STY\n",
      "    prefix_reference: http://purl.bioontology.org/ontology/STY/\n",
      "  UMLSSG:\n",
      "    prefix_prefix: UMLSSG\n",
      "    prefix_reference: https://lhncbc.nlm.nih.gov/semanticnetwork/download/sg_archive/SemGroups-v04.txt\n",
      "  UniProtKB:\n",
      "    prefix_prefix: UniProtKB\n",
      "    prefix_reference: http://purl.uniprot.org/uniprot/\n",
      "  UNIPROT.ISOFORM:\n",
      "    prefix_prefix: UNIPROT.ISOFORM\n",
      "    prefix_reference: http://purl.uniprot.org/isoforms/\n",
      "  VANDF:\n",
      "    prefix_prefix: VANDF\n",
      "    prefix_reference: https://www.nlm.nih.gov/research/umls/sourcereleasedocs/current/VANDF/\n",
      "  UO-PROPERTY:\n",
      "    prefix_prefix: UO-PROPERTY\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/uo#\n",
      "  VMC:\n",
      "    prefix_prefix: VMC\n",
      "    prefix_reference: https://github.com/ga4gh/vr-spec/\n",
      "  WBls:\n",
      "    prefix_prefix: WBls\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/WBls_\n",
      "  WBbt:\n",
      "    prefix_prefix: WBbt\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/WBbt_\n",
      "  WBVocab:\n",
      "    prefix_prefix: WBVocab\n",
      "    prefix_reference: http://bio2rdf.org/wormbase_vocabulary\n",
      "  WIKIDATA:\n",
      "    prefix_prefix: WIKIDATA\n",
      "    prefix_reference: https://www.wikidata.org/entity/\n",
      "  WIKIDATA_PROPERTY:\n",
      "    prefix_prefix: WIKIDATA_PROPERTY\n",
      "    prefix_reference: https://www.wikidata.org/prop/\n",
      "  wgs:\n",
      "    prefix_prefix: wgs\n",
      "    prefix_reference: http://www.w3.org/2003/01/geo/wgs84_pos\n",
      "  XPO:\n",
      "    prefix_prefix: XPO\n",
      "    prefix_reference: http://purl.obolibrary.org/obo/XPO_\n",
      "  Xenbase:\n",
      "    prefix_prefix: Xenbase\n",
      "    prefix_reference: http://www.xenbase.org/gene/showgene.do?method=display&geneId=\n",
      "  PMC:\n",
      "    prefix_prefix: PMC\n",
      "    prefix_reference: http://europepmc.org/articles/PMC\n",
      "default_prefix: biolink-subset/\n",
      "default_range: string\n",
      "classes:\n",
      "  Gene:\n",
      "    name: Gene\n",
      "    id_prefixes:\n",
      "    - NCBIGene\n",
      "    - ENSEMBL\n",
      "    - HGNC\n",
      "    - MGI\n",
      "    - ZFIN\n",
      "    - dictyBase\n",
      "    - WB\n",
      "    - WormBase\n",
      "    - FB\n",
      "    - RGD\n",
      "    - SGD\n",
      "    - PomBase\n",
      "    - OMIM\n",
      "    - KEGG.GENES\n",
      "    - UMLS\n",
      "    - Xenbase\n",
      "    - AspGD\n",
      "    - PHARMGKB.GENE\n",
      "    description: A region (or regions) that includes all of the sequence elements\n",
      "      necessary to encode a functional transcript. A gene locus may include regulatory\n",
      "      regions, transcribed regions and/or other functional sequence regions.\n",
      "    in_subset:\n",
      "    - translator_minimal\n",
      "    - model_organism_database\n",
      "    exact_mappings:\n",
      "    - SO:0000704\n",
      "    - SIO:010035\n",
      "    - WIKIDATA:Q7187\n",
      "    - dcid:Gene\n",
      "    narrow_mappings:\n",
      "    - bioschemas:gene\n",
      "    broad_mappings:\n",
      "    - NCIT:C45822\n",
      "    attributes:\n",
      "      symbol:\n",
      "        name: symbol\n",
      "        description: Symbol for a particular thing\n",
      "        exact_mappings:\n",
      "        - AGRKB:symbol\n",
      "        - gpi:DB_Object_Symbol\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - gene\n",
      "      xref:\n",
      "        name: xref\n",
      "        description: A database cross reference or alternative identifier for a NamedThing\n",
      "          or edge between two NamedThings.  This property should point to a database\n",
      "          record or webpage that supports the existence of the edge, or gives more\n",
      "          detail about the edge. This property can be used on a node or edge to provide\n",
      "          multiple URIs or CURIE cross references.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - dbxref\n",
      "        - Dbxref\n",
      "        - DbXref\n",
      "        - record_url\n",
      "        - source_record_urls\n",
      "        narrow_mappings:\n",
      "        - gff3:Dbxref\n",
      "        - gpi:DB_Xrefs\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - publication\n",
      "        - retrieval source\n",
      "        - gene\n",
      "        - gene product mixin\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      has_biological_sequence:\n",
      "        name: has_biological_sequence\n",
      "        description: connects a genomic feature to its sequence\n",
      "        is_a: node property\n",
      "        domain_of:\n",
      "        - genomic entity\n",
      "        - epigenomic entity\n",
      "        range: biological sequence\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      in_taxon:\n",
      "        name: in_taxon\n",
      "        annotations:\n",
      "          canonical_predicate:\n",
      "            tag: canonical_predicate\n",
      "            value: true\n",
      "        description: connects an entity to its taxonomic classification. Only certain\n",
      "          kinds of entities can be taxonomically classified; see 'thing with taxon'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - instance of\n",
      "        - is organism source of gene product\n",
      "        - organism has gene\n",
      "        - gene found in organism\n",
      "        - gene product has organism source\n",
      "        exact_mappings:\n",
      "        - RO:0002162\n",
      "        - WIKIDATA_PROPERTY:P703\n",
      "        narrow_mappings:\n",
      "        - RO:0002160\n",
      "        is_a: related to at instance level\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: organism taxon\n",
      "      in_taxon_label:\n",
      "        name: in_taxon_label\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: The human readable scientific name for the taxon of the entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P225\n",
      "        is_a: node property\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: label type\n",
      "      provided_by:\n",
      "        name: provided_by\n",
      "        description: The value in this node property represents the knowledge provider\n",
      "          that created or assembled the node and all of its attributes.  Used internally\n",
      "          to represent how a particular node made its way into a knowledge provider\n",
      "          or graph.\n",
      "        is_a: node property\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        multivalued: true\n",
      "      full_name:\n",
      "        name: full_name\n",
      "        description: a long-form human readable name for a thing\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        range: label type\n",
      "      synonym:\n",
      "        name: synonym\n",
      "        description: Alternate human-readable names for a thing\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - alias\n",
      "        narrow_mappings:\n",
      "        - skos:altLabel\n",
      "        - gff3:Alias\n",
      "        - AGRKB:synonyms\n",
      "        - gpi:DB_Object_Synonyms\n",
      "        - HANCESTRO:0330\n",
      "        - IAO:0000136\n",
      "        - RXNORM:has_tradename\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - gene product mixin\n",
      "        range: label type\n",
      "        multivalued: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "  Disease:\n",
      "    name: Disease\n",
      "    id_prefixes:\n",
      "    - MONDO\n",
      "    - DOID\n",
      "    - OMIM\n",
      "    - OMIM.PS\n",
      "    - orphanet\n",
      "    - EFO\n",
      "    - UMLS\n",
      "    - MESH\n",
      "    - MEDDRA\n",
      "    - NCIT\n",
      "    - SNOMEDCT\n",
      "    - medgen\n",
      "    - ICD10\n",
      "    - ICD9\n",
      "    - KEGG.DISEASE\n",
      "    - HP\n",
      "    - MP\n",
      "    - PHARMGKB.DISEASE\n",
      "    description: A disorder of structure or function, especially one that produces\n",
      "      specific signs, phenotypes or symptoms or that affects a specific location and\n",
      "      is not simply a direct result of physical injury.  A disposition to undergo\n",
      "      pathological processes that exists in an organism because of one or more disorders\n",
      "      in that organism.\n",
      "    in_subset:\n",
      "    - model_organism_database\n",
      "    - translator_minimal\n",
      "    aliases:\n",
      "    - condition\n",
      "    - disorder\n",
      "    - medical condition\n",
      "    exact_mappings:\n",
      "    - MONDO:0000001\n",
      "    - DOID:4\n",
      "    - NCIT:C2991\n",
      "    - WIKIDATA:Q12136\n",
      "    - SIO:010299\n",
      "    - UMLSSG:DISO\n",
      "    - STY:T047\n",
      "    - dcid:Disease\n",
      "    narrow_mappings:\n",
      "    - STY:T019\n",
      "    - STY:T020\n",
      "    - STY:T048\n",
      "    - STY:T049\n",
      "    - STY:T191\n",
      "    - MONDO:0042489\n",
      "    attributes:\n",
      "      in_taxon:\n",
      "        name: in_taxon\n",
      "        annotations:\n",
      "          canonical_predicate:\n",
      "            tag: canonical_predicate\n",
      "            value: true\n",
      "        description: connects an entity to its taxonomic classification. Only certain\n",
      "          kinds of entities can be taxonomically classified; see 'thing with taxon'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - instance of\n",
      "        - is organism source of gene product\n",
      "        - organism has gene\n",
      "        - gene found in organism\n",
      "        - gene product has organism source\n",
      "        exact_mappings:\n",
      "        - RO:0002162\n",
      "        - WIKIDATA_PROPERTY:P703\n",
      "        narrow_mappings:\n",
      "        - RO:0002160\n",
      "        is_a: related to at instance level\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: organism taxon\n",
      "      in_taxon_label:\n",
      "        name: in_taxon_label\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: The human readable scientific name for the taxon of the entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P225\n",
      "        is_a: node property\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: label type\n",
      "      provided_by:\n",
      "        name: provided_by\n",
      "        description: The value in this node property represents the knowledge provider\n",
      "          that created or assembled the node and all of its attributes.  Used internally\n",
      "          to represent how a particular node made its way into a knowledge provider\n",
      "          or graph.\n",
      "        is_a: node property\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        multivalued: true\n",
      "      xref:\n",
      "        name: xref\n",
      "        description: A database cross reference or alternative identifier for a NamedThing\n",
      "          or edge between two NamedThings.  This property should point to a database\n",
      "          record or webpage that supports the existence of the edge, or gives more\n",
      "          detail about the edge. This property can be used on a node or edge to provide\n",
      "          multiple URIs or CURIE cross references.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - dbxref\n",
      "        - Dbxref\n",
      "        - DbXref\n",
      "        - record_url\n",
      "        - source_record_urls\n",
      "        narrow_mappings:\n",
      "        - gff3:Dbxref\n",
      "        - gpi:DB_Xrefs\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - publication\n",
      "        - retrieval source\n",
      "        - gene\n",
      "        - gene product mixin\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      full_name:\n",
      "        name: full_name\n",
      "        description: a long-form human readable name for a thing\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        range: label type\n",
      "      synonym:\n",
      "        name: synonym\n",
      "        description: Alternate human-readable names for a thing\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - alias\n",
      "        narrow_mappings:\n",
      "        - skos:altLabel\n",
      "        - gff3:Alias\n",
      "        - AGRKB:synonyms\n",
      "        - gpi:DB_Object_Synonyms\n",
      "        - HANCESTRO:0330\n",
      "        - IAO:0000136\n",
      "        - RXNORM:has_tradename\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - gene product mixin\n",
      "        range: label type\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "  CaseToPhenotypicFeatureAssociation:\n",
      "    name: CaseToPhenotypicFeatureAssociation\n",
      "    description: An association between a case (e.g. individual patient) and a phenotypic\n",
      "      feature in which the individual has or has had the phenotype.\n",
      "    attributes:\n",
      "      subject:\n",
      "        name: subject\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation subject\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with outgoing relationship\n",
      "        description: connects an association to the subject of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedSource\n",
      "        - OBAN:association_has_subject\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:subject\n",
      "        range: named thing\n",
      "        required: true\n",
      "      predicate:\n",
      "        name: predicate\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation predicate\n",
      "          translator:\n",
      "            local_name_source: translator\n",
      "            local_name_value: predicate\n",
      "        description: A high-level grouping for the relationship type. AKA minimal\n",
      "          predicate. This is analogous to category for nodes.\n",
      "        notes:\n",
      "        - Has a value from the Biolink related_to hierarchy. In RDF,  this corresponds\n",
      "          to rdf:predicate and in Neo4j this corresponds to the relationship type.\n",
      "          The convention is for an edge label in snake_case form. For example, biolink:related_to,\n",
      "          biolink:causes, biolink:treats\n",
      "        exact_mappings:\n",
      "        - owl:annotatedProperty\n",
      "        - OBAN:association_has_predicate\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:predicate\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - association\n",
      "        - cell line to entity association mixin\n",
      "        - chemical entity to entity association mixin\n",
      "        - drug to entity association mixin\n",
      "        - chemical to entity association mixin\n",
      "        - case to entity association mixin\n",
      "        - chemical to chemical association\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - material sample to entity association mixin\n",
      "        - material sample derivation association\n",
      "        - disease to entity association mixin\n",
      "        - entity to exposure event association mixin\n",
      "        - entity to outcome association mixin\n",
      "        - frequency qualifier mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - disease or phenotypic feature to entity association mixin\n",
      "        - entity to disease or phenotypic feature association mixin\n",
      "        - genotype to entity association mixin\n",
      "        - gene to entity association mixin\n",
      "        - variant to entity association mixin\n",
      "        - model to disease association mixin\n",
      "        - macromolecular machine to entity association mixin\n",
      "        - organism taxon to entity association\n",
      "        range: predicate type\n",
      "        required: true\n",
      "      object:\n",
      "        name: object\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: descriptor\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with incoming relationship\n",
      "        description: connects an association to the object of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedTarget\n",
      "        - OBAN:association_has_object\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:object\n",
      "        range: named thing\n",
      "        required: true\n",
      "      sex_qualifier:\n",
      "        name: sex_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state whether\n",
      "          the association is specific to a particular sex.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - phenotypic feature to entity association mixin\n",
      "        range: biological sex\n",
      "      disease_context_qualifier:\n",
      "        name: disease_context_qualifier\n",
      "        description: A context qualifier representing a disease or condition in which\n",
      "          a relationship expressed in an association took place.\n",
      "        examples:\n",
      "        - value: MONDO:0004979\n",
      "        - value: MONDO:0005148\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: context qualifier\n",
      "        domain_of:\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: disease\n",
      "      subject_specialization_qualifier:\n",
      "        name: subject_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      object_specialization_qualifier:\n",
      "        name: object_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      anatomical_context_qualifier:\n",
      "        name: anatomical_context_qualifier\n",
      "        description: A statement qualifier representing an anatomical location where\n",
      "          an relationship expressed in an association took place (can be a tissue,\n",
      "          cell type, or sub-cellular location).\n",
      "        notes:\n",
      "        - Anatomical context values can be any term from UBERON. For example, the\n",
      "          context qualifier ‘cerebral cortext’ combines with a core concept of ‘neuron’\n",
      "          to express the composed concept ‘neuron in the cerebral cortext’. The species_context_qualifier\n",
      "          applies taxonomic context, e.g. species-specific molecular activity.  Ontology\n",
      "          CURIEs are expected as values here, the examples below are intended to help\n",
      "          clarify the content of the CURIEs.\n",
      "        examples:\n",
      "        - value: blood\n",
      "        - value: cerebral cortext\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: statement qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical gene interaction association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to phenotypic feature association mixin\n",
      "      negated:\n",
      "        name: negated\n",
      "        description: if set to true, then the association is negated i.e. is not true\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: boolean\n",
      "      qualifier:\n",
      "        name: qualifier\n",
      "        description: grouping slot for all qualifiers on an edge.  useful for testing\n",
      "          compliance with association classes\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      qualifiers:\n",
      "        name: qualifiers\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation qualifier\n",
      "        description: connects an association to qualifiers that modify or qualify\n",
      "          the meaning of that association\n",
      "        deprecated: 'True'\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      publications:\n",
      "        name: publications\n",
      "        description: One or more publications that report the statement expressed\n",
      "          in an Association, or provide information used as evidence supporting this\n",
      "          statement.\n",
      "        comments:\n",
      "        - The notion of a ‘Publication’ is considered broadly to include any document\n",
      "          made available for public consumption. It covers journal issues, individual\n",
      "          articles, and books - and also things like article pre-prints, white papers,\n",
      "          patents, drug labels, web pages, protocol documents, etc.\n",
      "        aliases:\n",
      "        - supporting publications\n",
      "        - supporting documents\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: publication\n",
      "        multivalued: true\n",
      "      has_evidence:\n",
      "        name: has_evidence\n",
      "        description: connects an association to an instance of supporting evidence\n",
      "        exact_mappings:\n",
      "        - RO:0002558\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: evidence type\n",
      "        multivalued: true\n",
      "      knowledge_source:\n",
      "        name: knowledge_source\n",
      "        description: An Information Resource from which the knowledge expressed in\n",
      "          an Association was retrieved, directly or indirectly. This can be any resource\n",
      "          through which the knowledge passed on its way to its currently serialized\n",
      "          form. In practice, implementers should use one of the more specific subtypes\n",
      "          of this generic property.\n",
      "        close_mappings:\n",
      "        - pav:providedBy\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      primary_knowledge_source:\n",
      "        name: primary_knowledge_source\n",
      "        description: The most upstream source of the knowledge expressed in an Association\n",
      "          that an implementer can identify.  Performing a rigorous analysis of upstream\n",
      "          data providers is expected; every effort is made to catalog the most upstream\n",
      "          source of data in this property.  Only one data source should be declared\n",
      "          primary in any association.  \"aggregator knowledge source\" can be used to\n",
      "          capture non-primary sources.\n",
      "        notes:\n",
      "        - 'For example: a single ChemicalToGene Edge originally curated by ClinicalTrials.org,\n",
      "          is aggregated by ChEMBL, then incorporated into the MolePro KP, then sent\n",
      "          via TRAPI message to the ARAGORN ARA, and finally sent to the NCATS ARS.\n",
      "          The retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   MolePro  --retrieved_from-->\n",
      "          ChEMBL --retrieved_from-->  ClinicalTrials.gov The \"primary knowledge source\"\n",
      "          for this edge is \"infores:clinical-trials-gov\".  \"infores:chembl\" and \"infores:molecular_data_provider\"\n",
      "          are listed in the \"aggregator knowledge source\" property.'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: false\n",
      "      aggregator_knowledge_source:\n",
      "        name: aggregator_knowledge_source\n",
      "        description: An intermediate aggregator resource from which knowledge expressed\n",
      "          in an Association was retrieved downstream of the original source, on its\n",
      "          path to its current serialized form.\n",
      "        notes:\n",
      "        - 'For example, in this Feature Variable Association Edge generated by the\n",
      "          Exposure Agent’s ICEES KP, through statistical analysis of clinical and\n",
      "          environmental data supplied by the UNC Clinical Data Warehouse, the Edge\n",
      "          is passed to the Ranking Agent’s ARAGORN ARA, and then on to the ARS. The\n",
      "          retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   ICEES\n",
      "          --supporting_data_from-->  UNC Data Warehouse This example illustrates how\n",
      "          to represent the source provenance of KP-generated knowledge, including\n",
      "          the source of data from which the knowledge was derived. The \"primary knowledge\n",
      "          source\" for this edge is \"infores:icees-asthma\". A \"supporting data source\"\n",
      "          for this KP- generated knowledge is \"infores:unc-cdw-health.\"  The \"aggregator\n",
      "          knowledge source\" for this data is \"infores:aragorn-ara\"'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      knowledge_level:\n",
      "        name: knowledge_level\n",
      "        description: Describes the level of knowledge expressed in a statement, based\n",
      "          on the reasoning or analysis methods used to generate the statement, or\n",
      "          the scope or specificity of what the statement expresses to be true.\n",
      "        notes:\n",
      "        - The notion of a 'level' of knowledge can in one sense relate to the strength  of\n",
      "          a statement - i.e. how confident we are that it says something true about\n",
      "          our domain of discourse. Here, we can generally consider Assertions to be  stronger\n",
      "          than Entailments to be stronger than Predictions.  But in another sense,\n",
      "          'level' of knowledge can refer to the scope or specificity of what a statement\n",
      "          expresses - on a spectrum from context-specific results of a data analysis,\n",
      "          to generalized assertions of knowledge or fact. Here, Statistical  Associations\n",
      "          and  Observations represent more foundational statements that are only  slightly\n",
      "          removed from the data on which they are based (the former reporting the  direct\n",
      "          results of  an analysis in terms of correlations between variables in the\n",
      "          data, and the latter describing phenomena that were observed/reported to\n",
      "          have occurred).\n",
      "        examples:\n",
      "        - value: knowledge_assertion\n",
      "        - value: prediction\n",
      "        - value: statistical_association\n",
      "        aliases:\n",
      "        - knowledge type\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: KnowledgeLevelEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      agent_type:\n",
      "        name: agent_type\n",
      "        description: Describes the high-level category of agent who originally generated\n",
      "          a  statement of knowledge or other type of information.\n",
      "        notes:\n",
      "        - Note that this property indicates the type of agent who produced a  final\n",
      "          statement of knowledge, which is often different from the  agent oragents\n",
      "          who produced information used as evidence to  support generation of this\n",
      "          knowledge. For example, if a human curator concludes that a particular gene\n",
      "          variant causes a medical condition - based on their interpretation of information\n",
      "          produced by computational  modeling tools, automated data analysis pipelines,\n",
      "          and robotic laboratory assay systems - the agent_type for this statement\n",
      "          is 'manual agent' -  despite all of the evidence being created by automated\n",
      "          agents. But if any of these systems is programmed to generate knowledge\n",
      "          statements  directly and without human assistance, the statement would be\n",
      "          attributed  to an 'automated_agent'.\n",
      "        examples:\n",
      "        - value: manual_agent\n",
      "        - value: automated_agent\n",
      "        - value: computational_model\n",
      "        - value: text_mining_agent\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: AgentTypeEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      timepoint:\n",
      "        name: timepoint\n",
      "        description: a point in time\n",
      "        aliases:\n",
      "        - duration\n",
      "        domain_of:\n",
      "        - geographic location at time\n",
      "        - exposure event\n",
      "        - association\n",
      "        range: time type\n",
      "      original_subject:\n",
      "        name: original_subject\n",
      "        description: used to hold the original subject of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      original_predicate:\n",
      "        name: original_predicate\n",
      "        id_prefixes:\n",
      "        - RO\n",
      "        - BSPO\n",
      "        - SIO\n",
      "        description: used to hold the original relation/predicate that an external\n",
      "          knowledge source uses before transformation to match the biolink-model specification.\n",
      "        aliases:\n",
      "        - original relation\n",
      "        - relation\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: uriorcurie\n",
      "      original_object:\n",
      "        name: original_object\n",
      "        description: used to hold the original object of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      subject_category:\n",
      "        name: subject_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Gene\n",
      "          description: The subject category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Gene'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      object_category:\n",
      "        name: object_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Disease\n",
      "          description: The object category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Disease'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      subject_closure:\n",
      "        name: subject_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      object_closure:\n",
      "        name: object_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''MONDO:0000167'', ''MONDO:0005395'']'\n",
      "          description: 'The object closure of the association between the gene ''BRCA1''\n",
      "            and the disease ''breast cancer'' is the set of all diseases that are\n",
      "            ancestors of ''breast cancer'' in the MONDO ontology.  Note: typically\n",
      "            the \"subclass of\" and \"part of\" relations are used to construct the closure,\n",
      "            but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      subject_category_closure:\n",
      "        name: subject_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject category closure of an association.\n",
      "          This is a denormalized field used primarily in the SQL serialization of\n",
      "          a knowledge graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Gene\", \"biolink:NamedThing'']'\n",
      "          description: 'The subject category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Gene'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      object_category_closure:\n",
      "        name: object_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object category closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Disease\", \"biolink:NamedThing'']'\n",
      "          description: 'The object category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Disease'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      subject_namespace:\n",
      "        name: subject_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: NCBIGene\n",
      "          description: The subject namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'NCBIGene'.\n",
      "        aliases:\n",
      "        - subject prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      object_namespace:\n",
      "        name: object_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: MONDO\n",
      "          description: The object namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'MONDO'.\n",
      "        aliases:\n",
      "        - object prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      subject_label_closure:\n",
      "        name: subject_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''BRACA1'']'\n",
      "          description: The subject label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'BRCA1' in the biolink model.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      object_label_closure:\n",
      "        name: object_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: breast cancer\n",
      "          description: The object label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'breast cancer' in the biolink model.\n",
      "        - value: cancer\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      retrieval_source_ids:\n",
      "        name: retrieval_source_ids\n",
      "        description: A list of retrieval sources that served as a source of knowledge\n",
      "          expressed in an Edge, or a source of data used to generate this knowledge.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: retrieval source\n",
      "        multivalued: true\n",
      "      p_value:\n",
      "        name: p_value\n",
      "        description: A quantitative confidence value that represents the probability\n",
      "          of obtaining a result at least as extreme as that actually obtained, assuming\n",
      "          that the actual value was the result of chance alone.\n",
      "        aliases:\n",
      "        - unadjusted p value\n",
      "        exact_mappings:\n",
      "        - OBI:0000175\n",
      "        - NCIT:C44185\n",
      "        - EDAM-DATA:1669\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      adjusted_p_value:\n",
      "        name: adjusted_p_value\n",
      "        description: The adjusted p-value is the probability of obtaining test results\n",
      "          at least as extreme as the results actually observed, under the assumption\n",
      "          that the null hypothesis is correct, adjusted for multiple comparisons.\n",
      "          P is always italicized and capitalized. The actual P value* should be expressed\n",
      "          (P=. 04) rather than expressing a statement of inequality (P<. 05), unless\n",
      "          P<.\n",
      "        is_a: p value\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      has_supporting_studies:\n",
      "        name: has_supporting_studies\n",
      "        description: A study that produced information used as evidence to generate\n",
      "          the knowledge expressed in an Association.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: study\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "      has_count:\n",
      "        name: has_count\n",
      "        description: number of things with a particular property\n",
      "        exact_mappings:\n",
      "        - LOINC:has_count\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_total:\n",
      "        name: has_total\n",
      "        description: total number of things in a particular reference set\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_quotient:\n",
      "        name: has_quotient\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      has_percentage:\n",
      "        name: has_percentage\n",
      "        description: equivalent to has quotient multiplied by 100\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      subject_aspect_qualifier:\n",
      "        name: subject_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the subject of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "      subject_direction_qualifier:\n",
      "        name: subject_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the subject\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        range: DirectionQualifierEnum\n",
      "      object_aspect_qualifier:\n",
      "        name: object_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the object of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      object_direction_qualifier:\n",
      "        name: object_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the object\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "        - chemical entity or gene or gene product regulates gene association\n",
      "        range: DirectionQualifierEnum\n",
      "      qualified_predicate:\n",
      "        name: qualified_predicate\n",
      "        description: Predicate to be used in an association when subject and object\n",
      "          qualifiers are present and the full reading of the statement requires a\n",
      "          qualification to the predicate in use in order to refine or increase the\n",
      "          specificity of the full statement reading.  This qualifier holds a relationship\n",
      "          to be used instead of that expressed by the primary predicate, in a ‘full\n",
      "          statement’ reading of the association, where qualifier-based semantics are\n",
      "          included.  This is necessary only in cases where the primary predicate does\n",
      "          not work in a full statement reading.\n",
      "        notes:\n",
      "        - 'to express the statement that “Chemical X causes increased expression of\n",
      "          Gene Y”, the core triple is read using the fields subject:ChemX, predicate:affects,\n",
      "          object:GeneY . . . and the full statement is read using the fields subject:ChemX,\n",
      "          qualified_predicate:causes, object:GeneY, object_aspect: expression, object_direction:increased.\n",
      "          The predicate ‘affects’ is needed for the core triple reading, but does\n",
      "          not make sense in the full statement reading  (because “Chemical X affects\n",
      "          increased expression of Gene Y'''' is not what we mean to say here: it causes\n",
      "          increased expression of Gene Y)'\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      frequency_qualifier:\n",
      "        name: frequency_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state how frequent\n",
      "          the phenotype is observed in the subject\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - frequency qualifier mixin\n",
      "        range: frequency value\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  GeneToDiseaseAssociation:\n",
      "    name: GeneToDiseaseAssociation\n",
      "    comments:\n",
      "    - NCIT:R176 refers to the inverse relationship\n",
      "    exact_mappings:\n",
      "    - SIO:000983\n",
      "    close_mappings:\n",
      "    - dcid:DiseaseGeneAssociation\n",
      "    attributes:\n",
      "      subject:\n",
      "        name: subject\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation subject\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with outgoing relationship\n",
      "        description: connects an association to the subject of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedSource\n",
      "        - OBAN:association_has_subject\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:subject\n",
      "        range: named thing\n",
      "        required: true\n",
      "      predicate:\n",
      "        name: predicate\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation predicate\n",
      "          translator:\n",
      "            local_name_source: translator\n",
      "            local_name_value: predicate\n",
      "        description: A high-level grouping for the relationship type. AKA minimal\n",
      "          predicate. This is analogous to category for nodes.\n",
      "        notes:\n",
      "        - Has a value from the Biolink related_to hierarchy. In RDF,  this corresponds\n",
      "          to rdf:predicate and in Neo4j this corresponds to the relationship type.\n",
      "          The convention is for an edge label in snake_case form. For example, biolink:related_to,\n",
      "          biolink:causes, biolink:treats\n",
      "        exact_mappings:\n",
      "        - owl:annotatedProperty\n",
      "        - OBAN:association_has_predicate\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:predicate\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - association\n",
      "        - cell line to entity association mixin\n",
      "        - chemical entity to entity association mixin\n",
      "        - drug to entity association mixin\n",
      "        - chemical to entity association mixin\n",
      "        - case to entity association mixin\n",
      "        - chemical to chemical association\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - material sample to entity association mixin\n",
      "        - material sample derivation association\n",
      "        - disease to entity association mixin\n",
      "        - entity to exposure event association mixin\n",
      "        - entity to outcome association mixin\n",
      "        - frequency qualifier mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - disease or phenotypic feature to entity association mixin\n",
      "        - entity to disease or phenotypic feature association mixin\n",
      "        - genotype to entity association mixin\n",
      "        - gene to entity association mixin\n",
      "        - variant to entity association mixin\n",
      "        - model to disease association mixin\n",
      "        - macromolecular machine to entity association mixin\n",
      "        - organism taxon to entity association\n",
      "        range: predicate type\n",
      "        required: true\n",
      "      object:\n",
      "        name: object\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: descriptor\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with incoming relationship\n",
      "        description: connects an association to the object of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedTarget\n",
      "        - OBAN:association_has_object\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:object\n",
      "        range: named thing\n",
      "        required: true\n",
      "      subject_aspect_qualifier:\n",
      "        name: subject_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the subject of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "      object_direction_qualifier:\n",
      "        name: object_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the object\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "        - chemical entity or gene or gene product regulates gene association\n",
      "        range: DirectionQualifierEnum\n",
      "      sex_qualifier:\n",
      "        name: sex_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state whether\n",
      "          the association is specific to a particular sex.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - phenotypic feature to entity association mixin\n",
      "        range: biological sex\n",
      "      disease_context_qualifier:\n",
      "        name: disease_context_qualifier\n",
      "        description: A context qualifier representing a disease or condition in which\n",
      "          a relationship expressed in an association took place.\n",
      "        examples:\n",
      "        - value: MONDO:0004979\n",
      "        - value: MONDO:0005148\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: context qualifier\n",
      "        domain_of:\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: disease\n",
      "      subject_specialization_qualifier:\n",
      "        name: subject_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      object_specialization_qualifier:\n",
      "        name: object_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      anatomical_context_qualifier:\n",
      "        name: anatomical_context_qualifier\n",
      "        description: A statement qualifier representing an anatomical location where\n",
      "          an relationship expressed in an association took place (can be a tissue,\n",
      "          cell type, or sub-cellular location).\n",
      "        notes:\n",
      "        - Anatomical context values can be any term from UBERON. For example, the\n",
      "          context qualifier ‘cerebral cortext’ combines with a core concept of ‘neuron’\n",
      "          to express the composed concept ‘neuron in the cerebral cortext’. The species_context_qualifier\n",
      "          applies taxonomic context, e.g. species-specific molecular activity.  Ontology\n",
      "          CURIEs are expected as values here, the examples below are intended to help\n",
      "          clarify the content of the CURIEs.\n",
      "        examples:\n",
      "        - value: blood\n",
      "        - value: cerebral cortext\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: statement qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical gene interaction association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to phenotypic feature association mixin\n",
      "      negated:\n",
      "        name: negated\n",
      "        description: if set to true, then the association is negated i.e. is not true\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: boolean\n",
      "      qualifier:\n",
      "        name: qualifier\n",
      "        description: grouping slot for all qualifiers on an edge.  useful for testing\n",
      "          compliance with association classes\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      qualifiers:\n",
      "        name: qualifiers\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation qualifier\n",
      "        description: connects an association to qualifiers that modify or qualify\n",
      "          the meaning of that association\n",
      "        deprecated: 'True'\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      publications:\n",
      "        name: publications\n",
      "        description: One or more publications that report the statement expressed\n",
      "          in an Association, or provide information used as evidence supporting this\n",
      "          statement.\n",
      "        comments:\n",
      "        - The notion of a ‘Publication’ is considered broadly to include any document\n",
      "          made available for public consumption. It covers journal issues, individual\n",
      "          articles, and books - and also things like article pre-prints, white papers,\n",
      "          patents, drug labels, web pages, protocol documents, etc.\n",
      "        aliases:\n",
      "        - supporting publications\n",
      "        - supporting documents\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: publication\n",
      "        multivalued: true\n",
      "      has_evidence:\n",
      "        name: has_evidence\n",
      "        description: connects an association to an instance of supporting evidence\n",
      "        exact_mappings:\n",
      "        - RO:0002558\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: evidence type\n",
      "        multivalued: true\n",
      "      knowledge_source:\n",
      "        name: knowledge_source\n",
      "        description: An Information Resource from which the knowledge expressed in\n",
      "          an Association was retrieved, directly or indirectly. This can be any resource\n",
      "          through which the knowledge passed on its way to its currently serialized\n",
      "          form. In practice, implementers should use one of the more specific subtypes\n",
      "          of this generic property.\n",
      "        close_mappings:\n",
      "        - pav:providedBy\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      primary_knowledge_source:\n",
      "        name: primary_knowledge_source\n",
      "        description: The most upstream source of the knowledge expressed in an Association\n",
      "          that an implementer can identify.  Performing a rigorous analysis of upstream\n",
      "          data providers is expected; every effort is made to catalog the most upstream\n",
      "          source of data in this property.  Only one data source should be declared\n",
      "          primary in any association.  \"aggregator knowledge source\" can be used to\n",
      "          capture non-primary sources.\n",
      "        notes:\n",
      "        - 'For example: a single ChemicalToGene Edge originally curated by ClinicalTrials.org,\n",
      "          is aggregated by ChEMBL, then incorporated into the MolePro KP, then sent\n",
      "          via TRAPI message to the ARAGORN ARA, and finally sent to the NCATS ARS.\n",
      "          The retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   MolePro  --retrieved_from-->\n",
      "          ChEMBL --retrieved_from-->  ClinicalTrials.gov The \"primary knowledge source\"\n",
      "          for this edge is \"infores:clinical-trials-gov\".  \"infores:chembl\" and \"infores:molecular_data_provider\"\n",
      "          are listed in the \"aggregator knowledge source\" property.'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: false\n",
      "      aggregator_knowledge_source:\n",
      "        name: aggregator_knowledge_source\n",
      "        description: An intermediate aggregator resource from which knowledge expressed\n",
      "          in an Association was retrieved downstream of the original source, on its\n",
      "          path to its current serialized form.\n",
      "        notes:\n",
      "        - 'For example, in this Feature Variable Association Edge generated by the\n",
      "          Exposure Agent’s ICEES KP, through statistical analysis of clinical and\n",
      "          environmental data supplied by the UNC Clinical Data Warehouse, the Edge\n",
      "          is passed to the Ranking Agent’s ARAGORN ARA, and then on to the ARS. The\n",
      "          retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   ICEES\n",
      "          --supporting_data_from-->  UNC Data Warehouse This example illustrates how\n",
      "          to represent the source provenance of KP-generated knowledge, including\n",
      "          the source of data from which the knowledge was derived. The \"primary knowledge\n",
      "          source\" for this edge is \"infores:icees-asthma\". A \"supporting data source\"\n",
      "          for this KP- generated knowledge is \"infores:unc-cdw-health.\"  The \"aggregator\n",
      "          knowledge source\" for this data is \"infores:aragorn-ara\"'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      knowledge_level:\n",
      "        name: knowledge_level\n",
      "        description: Describes the level of knowledge expressed in a statement, based\n",
      "          on the reasoning or analysis methods used to generate the statement, or\n",
      "          the scope or specificity of what the statement expresses to be true.\n",
      "        notes:\n",
      "        - The notion of a 'level' of knowledge can in one sense relate to the strength  of\n",
      "          a statement - i.e. how confident we are that it says something true about\n",
      "          our domain of discourse. Here, we can generally consider Assertions to be  stronger\n",
      "          than Entailments to be stronger than Predictions.  But in another sense,\n",
      "          'level' of knowledge can refer to the scope or specificity of what a statement\n",
      "          expresses - on a spectrum from context-specific results of a data analysis,\n",
      "          to generalized assertions of knowledge or fact. Here, Statistical  Associations\n",
      "          and  Observations represent more foundational statements that are only  slightly\n",
      "          removed from the data on which they are based (the former reporting the  direct\n",
      "          results of  an analysis in terms of correlations between variables in the\n",
      "          data, and the latter describing phenomena that were observed/reported to\n",
      "          have occurred).\n",
      "        examples:\n",
      "        - value: knowledge_assertion\n",
      "        - value: prediction\n",
      "        - value: statistical_association\n",
      "        aliases:\n",
      "        - knowledge type\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: KnowledgeLevelEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      agent_type:\n",
      "        name: agent_type\n",
      "        description: Describes the high-level category of agent who originally generated\n",
      "          a  statement of knowledge or other type of information.\n",
      "        notes:\n",
      "        - Note that this property indicates the type of agent who produced a  final\n",
      "          statement of knowledge, which is often different from the  agent oragents\n",
      "          who produced information used as evidence to  support generation of this\n",
      "          knowledge. For example, if a human curator concludes that a particular gene\n",
      "          variant causes a medical condition - based on their interpretation of information\n",
      "          produced by computational  modeling tools, automated data analysis pipelines,\n",
      "          and robotic laboratory assay systems - the agent_type for this statement\n",
      "          is 'manual agent' -  despite all of the evidence being created by automated\n",
      "          agents. But if any of these systems is programmed to generate knowledge\n",
      "          statements  directly and without human assistance, the statement would be\n",
      "          attributed  to an 'automated_agent'.\n",
      "        examples:\n",
      "        - value: manual_agent\n",
      "        - value: automated_agent\n",
      "        - value: computational_model\n",
      "        - value: text_mining_agent\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: AgentTypeEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      timepoint:\n",
      "        name: timepoint\n",
      "        description: a point in time\n",
      "        aliases:\n",
      "        - duration\n",
      "        domain_of:\n",
      "        - geographic location at time\n",
      "        - exposure event\n",
      "        - association\n",
      "        range: time type\n",
      "      original_subject:\n",
      "        name: original_subject\n",
      "        description: used to hold the original subject of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      original_predicate:\n",
      "        name: original_predicate\n",
      "        id_prefixes:\n",
      "        - RO\n",
      "        - BSPO\n",
      "        - SIO\n",
      "        description: used to hold the original relation/predicate that an external\n",
      "          knowledge source uses before transformation to match the biolink-model specification.\n",
      "        aliases:\n",
      "        - original relation\n",
      "        - relation\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: uriorcurie\n",
      "      original_object:\n",
      "        name: original_object\n",
      "        description: used to hold the original object of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      subject_category:\n",
      "        name: subject_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Gene\n",
      "          description: The subject category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Gene'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      object_category:\n",
      "        name: object_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Disease\n",
      "          description: The object category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Disease'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      subject_closure:\n",
      "        name: subject_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      object_closure:\n",
      "        name: object_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''MONDO:0000167'', ''MONDO:0005395'']'\n",
      "          description: 'The object closure of the association between the gene ''BRCA1''\n",
      "            and the disease ''breast cancer'' is the set of all diseases that are\n",
      "            ancestors of ''breast cancer'' in the MONDO ontology.  Note: typically\n",
      "            the \"subclass of\" and \"part of\" relations are used to construct the closure,\n",
      "            but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      subject_category_closure:\n",
      "        name: subject_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject category closure of an association.\n",
      "          This is a denormalized field used primarily in the SQL serialization of\n",
      "          a knowledge graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Gene\", \"biolink:NamedThing'']'\n",
      "          description: 'The subject category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Gene'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      object_category_closure:\n",
      "        name: object_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object category closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Disease\", \"biolink:NamedThing'']'\n",
      "          description: 'The object category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Disease'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      subject_namespace:\n",
      "        name: subject_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: NCBIGene\n",
      "          description: The subject namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'NCBIGene'.\n",
      "        aliases:\n",
      "        - subject prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      object_namespace:\n",
      "        name: object_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: MONDO\n",
      "          description: The object namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'MONDO'.\n",
      "        aliases:\n",
      "        - object prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      subject_label_closure:\n",
      "        name: subject_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''BRACA1'']'\n",
      "          description: The subject label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'BRCA1' in the biolink model.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      object_label_closure:\n",
      "        name: object_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: breast cancer\n",
      "          description: The object label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'breast cancer' in the biolink model.\n",
      "        - value: cancer\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      retrieval_source_ids:\n",
      "        name: retrieval_source_ids\n",
      "        description: A list of retrieval sources that served as a source of knowledge\n",
      "          expressed in an Edge, or a source of data used to generate this knowledge.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: retrieval source\n",
      "        multivalued: true\n",
      "      p_value:\n",
      "        name: p_value\n",
      "        description: A quantitative confidence value that represents the probability\n",
      "          of obtaining a result at least as extreme as that actually obtained, assuming\n",
      "          that the actual value was the result of chance alone.\n",
      "        aliases:\n",
      "        - unadjusted p value\n",
      "        exact_mappings:\n",
      "        - OBI:0000175\n",
      "        - NCIT:C44185\n",
      "        - EDAM-DATA:1669\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      adjusted_p_value:\n",
      "        name: adjusted_p_value\n",
      "        description: The adjusted p-value is the probability of obtaining test results\n",
      "          at least as extreme as the results actually observed, under the assumption\n",
      "          that the null hypothesis is correct, adjusted for multiple comparisons.\n",
      "          P is always italicized and capitalized. The actual P value* should be expressed\n",
      "          (P=. 04) rather than expressing a statement of inequality (P<. 05), unless\n",
      "          P<.\n",
      "        is_a: p value\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      has_supporting_studies:\n",
      "        name: has_supporting_studies\n",
      "        description: A study that produced information used as evidence to generate\n",
      "          the knowledge expressed in an Association.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: study\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "      has_count:\n",
      "        name: has_count\n",
      "        description: number of things with a particular property\n",
      "        exact_mappings:\n",
      "        - LOINC:has_count\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_total:\n",
      "        name: has_total\n",
      "        description: total number of things in a particular reference set\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_quotient:\n",
      "        name: has_quotient\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      has_percentage:\n",
      "        name: has_percentage\n",
      "        description: equivalent to has quotient multiplied by 100\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      subject_direction_qualifier:\n",
      "        name: subject_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the subject\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        range: DirectionQualifierEnum\n",
      "      object_aspect_qualifier:\n",
      "        name: object_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the object of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      qualified_predicate:\n",
      "        name: qualified_predicate\n",
      "        description: Predicate to be used in an association when subject and object\n",
      "          qualifiers are present and the full reading of the statement requires a\n",
      "          qualification to the predicate in use in order to refine or increase the\n",
      "          specificity of the full statement reading.  This qualifier holds a relationship\n",
      "          to be used instead of that expressed by the primary predicate, in a ‘full\n",
      "          statement’ reading of the association, where qualifier-based semantics are\n",
      "          included.  This is necessary only in cases where the primary predicate does\n",
      "          not work in a full statement reading.\n",
      "        notes:\n",
      "        - 'to express the statement that “Chemical X causes increased expression of\n",
      "          Gene Y”, the core triple is read using the fields subject:ChemX, predicate:affects,\n",
      "          object:GeneY . . . and the full statement is read using the fields subject:ChemX,\n",
      "          qualified_predicate:causes, object:GeneY, object_aspect: expression, object_direction:increased.\n",
      "          The predicate ‘affects’ is needed for the core triple reading, but does\n",
      "          not make sense in the full statement reading  (because “Chemical X affects\n",
      "          increased expression of Gene Y'''' is not what we mean to say here: it causes\n",
      "          increased expression of Gene Y)'\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      frequency_qualifier:\n",
      "        name: frequency_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state how frequent\n",
      "          the phenotype is observed in the subject\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - frequency qualifier mixin\n",
      "        range: frequency value\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  GeneToPhenotypicFeatureAssociation:\n",
      "    name: GeneToPhenotypicFeatureAssociation\n",
      "    exact_mappings:\n",
      "    - WBVocab:Gene-Phenotype-Association\n",
      "    attributes:\n",
      "      subject:\n",
      "        name: subject\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation subject\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with outgoing relationship\n",
      "        description: connects an association to the subject of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedSource\n",
      "        - OBAN:association_has_subject\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:subject\n",
      "        range: named thing\n",
      "        required: true\n",
      "      predicate:\n",
      "        name: predicate\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation predicate\n",
      "          translator:\n",
      "            local_name_source: translator\n",
      "            local_name_value: predicate\n",
      "        description: A high-level grouping for the relationship type. AKA minimal\n",
      "          predicate. This is analogous to category for nodes.\n",
      "        notes:\n",
      "        - Has a value from the Biolink related_to hierarchy. In RDF,  this corresponds\n",
      "          to rdf:predicate and in Neo4j this corresponds to the relationship type.\n",
      "          The convention is for an edge label in snake_case form. For example, biolink:related_to,\n",
      "          biolink:causes, biolink:treats\n",
      "        exact_mappings:\n",
      "        - owl:annotatedProperty\n",
      "        - OBAN:association_has_predicate\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:predicate\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - association\n",
      "        - cell line to entity association mixin\n",
      "        - chemical entity to entity association mixin\n",
      "        - drug to entity association mixin\n",
      "        - chemical to entity association mixin\n",
      "        - case to entity association mixin\n",
      "        - chemical to chemical association\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - material sample to entity association mixin\n",
      "        - material sample derivation association\n",
      "        - disease to entity association mixin\n",
      "        - entity to exposure event association mixin\n",
      "        - entity to outcome association mixin\n",
      "        - frequency qualifier mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - disease or phenotypic feature to entity association mixin\n",
      "        - entity to disease or phenotypic feature association mixin\n",
      "        - genotype to entity association mixin\n",
      "        - gene to entity association mixin\n",
      "        - variant to entity association mixin\n",
      "        - model to disease association mixin\n",
      "        - macromolecular machine to entity association mixin\n",
      "        - organism taxon to entity association\n",
      "        range: predicate type\n",
      "        required: true\n",
      "      object:\n",
      "        name: object\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: descriptor\n",
      "          neo4j:\n",
      "            local_name_source: neo4j\n",
      "            local_name_value: node with incoming relationship\n",
      "        description: connects an association to the object of the association. For\n",
      "          example, in a gene-to-phenotype association, the gene is subject and phenotype\n",
      "          is object.\n",
      "        exact_mappings:\n",
      "        - owl:annotatedTarget\n",
      "        - OBAN:association_has_object\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        slot_uri: rdf:object\n",
      "        range: named thing\n",
      "        required: true\n",
      "      sex_qualifier:\n",
      "        name: sex_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state whether\n",
      "          the association is specific to a particular sex.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        - phenotypic feature to entity association mixin\n",
      "        range: biological sex\n",
      "      disease_context_qualifier:\n",
      "        name: disease_context_qualifier\n",
      "        description: A context qualifier representing a disease or condition in which\n",
      "          a relationship expressed in an association took place.\n",
      "        examples:\n",
      "        - value: MONDO:0004979\n",
      "        - value: MONDO:0005148\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: context qualifier\n",
      "        domain_of:\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: disease\n",
      "      subject_specialization_qualifier:\n",
      "        name: subject_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      object_specialization_qualifier:\n",
      "        name: object_specialization_qualifier\n",
      "        description: A qualifier that composes with a core subject/object concept\n",
      "          to define a more specific version of the  subject concept, specifically\n",
      "          using an ontology term that is not a subclass or descendant of the core\n",
      "          concept  and in the vast majority of cases, is of a different ontological\n",
      "          namespace than the category or namespace of the subject identifier.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: specialization qualifier\n",
      "        domain_of:\n",
      "        - entity to phenotypic feature association mixin\n",
      "        range: uriorcurie\n",
      "      anatomical_context_qualifier:\n",
      "        name: anatomical_context_qualifier\n",
      "        description: A statement qualifier representing an anatomical location where\n",
      "          an relationship expressed in an association took place (can be a tissue,\n",
      "          cell type, or sub-cellular location).\n",
      "        notes:\n",
      "        - Anatomical context values can be any term from UBERON. For example, the\n",
      "          context qualifier ‘cerebral cortext’ combines with a core concept of ‘neuron’\n",
      "          to express the composed concept ‘neuron in the cerebral cortext’. The species_context_qualifier\n",
      "          applies taxonomic context, e.g. species-specific molecular activity.  Ontology\n",
      "          CURIEs are expected as values here, the examples below are intended to help\n",
      "          clarify the content of the CURIEs.\n",
      "        examples:\n",
      "        - value: blood\n",
      "        - value: cerebral cortext\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: statement qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical gene interaction association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to phenotypic feature association mixin\n",
      "      subject_aspect_qualifier:\n",
      "        name: subject_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the subject of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "      object_direction_qualifier:\n",
      "        name: object_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the object\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        - gene to disease or phenotypic feature association\n",
      "        - chemical entity or gene or gene product regulates gene association\n",
      "        range: DirectionQualifierEnum\n",
      "      negated:\n",
      "        name: negated\n",
      "        description: if set to true, then the association is negated i.e. is not true\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: boolean\n",
      "      qualifier:\n",
      "        name: qualifier\n",
      "        description: grouping slot for all qualifiers on an edge.  useful for testing\n",
      "          compliance with association classes\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      qualifiers:\n",
      "        name: qualifiers\n",
      "        local_names:\n",
      "          ga4gh:\n",
      "            local_name_source: ga4gh\n",
      "            local_name_value: annotation qualifier\n",
      "        description: connects an association to qualifiers that modify or qualify\n",
      "          the meaning of that association\n",
      "        deprecated: 'True'\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      publications:\n",
      "        name: publications\n",
      "        description: One or more publications that report the statement expressed\n",
      "          in an Association, or provide information used as evidence supporting this\n",
      "          statement.\n",
      "        comments:\n",
      "        - The notion of a ‘Publication’ is considered broadly to include any document\n",
      "          made available for public consumption. It covers journal issues, individual\n",
      "          articles, and books - and also things like article pre-prints, white papers,\n",
      "          patents, drug labels, web pages, protocol documents, etc.\n",
      "        aliases:\n",
      "        - supporting publications\n",
      "        - supporting documents\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: publication\n",
      "        multivalued: true\n",
      "      has_evidence:\n",
      "        name: has_evidence\n",
      "        description: connects an association to an instance of supporting evidence\n",
      "        exact_mappings:\n",
      "        - RO:0002558\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: evidence type\n",
      "        multivalued: true\n",
      "      knowledge_source:\n",
      "        name: knowledge_source\n",
      "        description: An Information Resource from which the knowledge expressed in\n",
      "          an Association was retrieved, directly or indirectly. This can be any resource\n",
      "          through which the knowledge passed on its way to its currently serialized\n",
      "          form. In practice, implementers should use one of the more specific subtypes\n",
      "          of this generic property.\n",
      "        close_mappings:\n",
      "        - pav:providedBy\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      primary_knowledge_source:\n",
      "        name: primary_knowledge_source\n",
      "        description: The most upstream source of the knowledge expressed in an Association\n",
      "          that an implementer can identify.  Performing a rigorous analysis of upstream\n",
      "          data providers is expected; every effort is made to catalog the most upstream\n",
      "          source of data in this property.  Only one data source should be declared\n",
      "          primary in any association.  \"aggregator knowledge source\" can be used to\n",
      "          capture non-primary sources.\n",
      "        notes:\n",
      "        - 'For example: a single ChemicalToGene Edge originally curated by ClinicalTrials.org,\n",
      "          is aggregated by ChEMBL, then incorporated into the MolePro KP, then sent\n",
      "          via TRAPI message to the ARAGORN ARA, and finally sent to the NCATS ARS.\n",
      "          The retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   MolePro  --retrieved_from-->\n",
      "          ChEMBL --retrieved_from-->  ClinicalTrials.gov The \"primary knowledge source\"\n",
      "          for this edge is \"infores:clinical-trials-gov\".  \"infores:chembl\" and \"infores:molecular_data_provider\"\n",
      "          are listed in the \"aggregator knowledge source\" property.'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: false\n",
      "      aggregator_knowledge_source:\n",
      "        name: aggregator_knowledge_source\n",
      "        description: An intermediate aggregator resource from which knowledge expressed\n",
      "          in an Association was retrieved downstream of the original source, on its\n",
      "          path to its current serialized form.\n",
      "        notes:\n",
      "        - 'For example, in this Feature Variable Association Edge generated by the\n",
      "          Exposure Agent’s ICEES KP, through statistical analysis of clinical and\n",
      "          environmental data supplied by the UNC Clinical Data Warehouse, the Edge\n",
      "          is passed to the Ranking Agent’s ARAGORN ARA, and then on to the ARS. The\n",
      "          retrieval path for this Edge is as follows: ARS--retrieved_from-->  ARAGORN  --retrieved_from-->   ICEES\n",
      "          --supporting_data_from-->  UNC Data Warehouse This example illustrates how\n",
      "          to represent the source provenance of KP-generated knowledge, including\n",
      "          the source of data from which the knowledge was derived. The \"primary knowledge\n",
      "          source\" for this edge is \"infores:icees-asthma\". A \"supporting data source\"\n",
      "          for this KP- generated knowledge is \"infores:unc-cdw-health.\"  The \"aggregator\n",
      "          knowledge source\" for this data is \"infores:aragorn-ara\"'\n",
      "        is_a: knowledge source\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      knowledge_level:\n",
      "        name: knowledge_level\n",
      "        description: Describes the level of knowledge expressed in a statement, based\n",
      "          on the reasoning or analysis methods used to generate the statement, or\n",
      "          the scope or specificity of what the statement expresses to be true.\n",
      "        notes:\n",
      "        - The notion of a 'level' of knowledge can in one sense relate to the strength  of\n",
      "          a statement - i.e. how confident we are that it says something true about\n",
      "          our domain of discourse. Here, we can generally consider Assertions to be  stronger\n",
      "          than Entailments to be stronger than Predictions.  But in another sense,\n",
      "          'level' of knowledge can refer to the scope or specificity of what a statement\n",
      "          expresses - on a spectrum from context-specific results of a data analysis,\n",
      "          to generalized assertions of knowledge or fact. Here, Statistical  Associations\n",
      "          and  Observations represent more foundational statements that are only  slightly\n",
      "          removed from the data on which they are based (the former reporting the  direct\n",
      "          results of  an analysis in terms of correlations between variables in the\n",
      "          data, and the latter describing phenomena that were observed/reported to\n",
      "          have occurred).\n",
      "        examples:\n",
      "        - value: knowledge_assertion\n",
      "        - value: prediction\n",
      "        - value: statistical_association\n",
      "        aliases:\n",
      "        - knowledge type\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: KnowledgeLevelEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      agent_type:\n",
      "        name: agent_type\n",
      "        description: Describes the high-level category of agent who originally generated\n",
      "          a  statement of knowledge or other type of information.\n",
      "        notes:\n",
      "        - Note that this property indicates the type of agent who produced a  final\n",
      "          statement of knowledge, which is often different from the  agent oragents\n",
      "          who produced information used as evidence to  support generation of this\n",
      "          knowledge. For example, if a human curator concludes that a particular gene\n",
      "          variant causes a medical condition - based on their interpretation of information\n",
      "          produced by computational  modeling tools, automated data analysis pipelines,\n",
      "          and robotic laboratory assay systems - the agent_type for this statement\n",
      "          is 'manual agent' -  despite all of the evidence being created by automated\n",
      "          agents. But if any of these systems is programmed to generate knowledge\n",
      "          statements  directly and without human assistance, the statement would be\n",
      "          attributed  to an 'automated_agent'.\n",
      "        examples:\n",
      "        - value: manual_agent\n",
      "        - value: automated_agent\n",
      "        - value: computational_model\n",
      "        - value: text_mining_agent\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: AgentTypeEnum\n",
      "        required: true\n",
      "        multivalued: false\n",
      "      timepoint:\n",
      "        name: timepoint\n",
      "        description: a point in time\n",
      "        aliases:\n",
      "        - duration\n",
      "        domain_of:\n",
      "        - geographic location at time\n",
      "        - exposure event\n",
      "        - association\n",
      "        range: time type\n",
      "      original_subject:\n",
      "        name: original_subject\n",
      "        description: used to hold the original subject of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      original_predicate:\n",
      "        name: original_predicate\n",
      "        id_prefixes:\n",
      "        - RO\n",
      "        - BSPO\n",
      "        - SIO\n",
      "        description: used to hold the original relation/predicate that an external\n",
      "          knowledge source uses before transformation to match the biolink-model specification.\n",
      "        aliases:\n",
      "        - original relation\n",
      "        - relation\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: uriorcurie\n",
      "      original_object:\n",
      "        name: original_object\n",
      "        description: used to hold the original object of a relation (or predicate)\n",
      "          that an external knowledge source uses before transformation to match the\n",
      "          biolink-model specification.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "      subject_category:\n",
      "        name: subject_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Gene\n",
      "          description: The subject category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Gene'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      object_category:\n",
      "        name: object_category\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the biolink class/category of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: biolink:Disease\n",
      "          description: The object category of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'biolink:Disease'.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: false\n",
      "      subject_closure:\n",
      "        name: subject_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      object_closure:\n",
      "        name: object_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object closure of an association. This is a\n",
      "          denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''MONDO:0000167'', ''MONDO:0005395'']'\n",
      "          description: 'The object closure of the association between the gene ''BRCA1''\n",
      "            and the disease ''breast cancer'' is the set of all diseases that are\n",
      "            ancestors of ''breast cancer'' in the MONDO ontology.  Note: typically\n",
      "            the \"subclass of\" and \"part of\" relations are used to construct the closure,\n",
      "            but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        multivalued: true\n",
      "      subject_category_closure:\n",
      "        name: subject_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject category closure of an association.\n",
      "          This is a denormalized field used primarily in the SQL serialization of\n",
      "          a knowledge graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Gene\", \"biolink:NamedThing'']'\n",
      "          description: 'The subject category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Gene'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      object_category_closure:\n",
      "        name: object_category_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object category closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''biolink:Disease\", \"biolink:NamedThing'']'\n",
      "          description: 'The object category closure of the association between the\n",
      "            gene ''BRCA1'' and the disease ''breast cancer'' is the set of all biolink\n",
      "            classes that are ancestors of ''biolink:Disease'' in the biolink model.  Note:\n",
      "            typically the \"subclass of\" and \"part of\" relations are used to construct\n",
      "            the closure, but other relations may be used as well.'\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: ontology class\n",
      "        multivalued: true\n",
      "      subject_namespace:\n",
      "        name: subject_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: NCBIGene\n",
      "          description: The subject namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'NCBIGene'.\n",
      "        aliases:\n",
      "        - subject prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      object_namespace:\n",
      "        name: object_namespace\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object namespace of an association. This is\n",
      "          a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: MONDO\n",
      "          description: The object namespace of the association between the gene 'BRCA1'\n",
      "            and the disease 'breast cancer' is 'MONDO'.\n",
      "        aliases:\n",
      "        - object prefix\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: false\n",
      "      subject_label_closure:\n",
      "        name: subject_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the subject label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: '[''BRACA1'']'\n",
      "          description: The subject label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'BRCA1' in the biolink model.\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      object_label_closure:\n",
      "        name: object_label_closure\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: Used to hold the object label closure of an association. This\n",
      "          is a denormalized field used primarily in the SQL serialization of a knowledge\n",
      "          graph via KGX.\n",
      "        examples:\n",
      "        - value: breast cancer\n",
      "          description: The object label closure of the association between the gene\n",
      "            'BRCA1' and the disease 'breast cancer' is the set of all labels that\n",
      "            are ancestors of 'breast cancer' in the biolink model.\n",
      "        - value: cancer\n",
      "        is_a: association slot\n",
      "        domain: association\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: string\n",
      "        multivalued: true\n",
      "      retrieval_source_ids:\n",
      "        name: retrieval_source_ids\n",
      "        description: A list of retrieval sources that served as a source of knowledge\n",
      "          expressed in an Edge, or a source of data used to generate this knowledge.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: retrieval source\n",
      "        multivalued: true\n",
      "      p_value:\n",
      "        name: p_value\n",
      "        description: A quantitative confidence value that represents the probability\n",
      "          of obtaining a result at least as extreme as that actually obtained, assuming\n",
      "          that the actual value was the result of chance alone.\n",
      "        aliases:\n",
      "        - unadjusted p value\n",
      "        exact_mappings:\n",
      "        - OBI:0000175\n",
      "        - NCIT:C44185\n",
      "        - EDAM-DATA:1669\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      adjusted_p_value:\n",
      "        name: adjusted_p_value\n",
      "        description: The adjusted p-value is the probability of obtaining test results\n",
      "          at least as extreme as the results actually observed, under the assumption\n",
      "          that the null hypothesis is correct, adjusted for multiple comparisons.\n",
      "          P is always italicized and capitalized. The actual P value* should be expressed\n",
      "          (P=. 04) rather than expressing a statement of inequality (P<. 05), unless\n",
      "          P<.\n",
      "        is_a: p value\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: float\n",
      "      has_supporting_studies:\n",
      "        name: has_supporting_studies\n",
      "        description: A study that produced information used as evidence to generate\n",
      "          the knowledge expressed in an Association.\n",
      "        is_a: association slot\n",
      "        domain_of:\n",
      "        - association\n",
      "        range: study\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "      has_count:\n",
      "        name: has_count\n",
      "        description: number of things with a particular property\n",
      "        exact_mappings:\n",
      "        - LOINC:has_count\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_total:\n",
      "        name: has_total\n",
      "        description: total number of things in a particular reference set\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: integer\n",
      "      has_quotient:\n",
      "        name: has_quotient\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      has_percentage:\n",
      "        name: has_percentage\n",
      "        description: equivalent to has quotient multiplied by 100\n",
      "        is_a: aggregate statistic\n",
      "        domain_of:\n",
      "        - frequency quantifier\n",
      "        range: double\n",
      "      subject_direction_qualifier:\n",
      "        name: subject_direction_qualifier\n",
      "        description: 'Composes with the core concept (+ aspect if provided) to describe\n",
      "          a change in its direction or degree. This qualifier qualifies the subject\n",
      "          of an association (aka: statement).'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: direction qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "        range: DirectionQualifierEnum\n",
      "      object_aspect_qualifier:\n",
      "        name: object_aspect_qualifier\n",
      "        description: 'Composes with the core concept to describe new concepts of a\n",
      "          different ontological type. e.g. a process in which the core concept participates,\n",
      "          a function/activity/role held by the core concept, or a characteristic/quality\n",
      "          that inheres in the core concept.  The purpose of the aspect slot is to\n",
      "          indicate what aspect is being affected in an ''affects'' association.  This\n",
      "          qualifier specifies a change in the object of an association (aka: statement).'\n",
      "        examples:\n",
      "        - value: stability\n",
      "        - value: abundance\n",
      "        - value: expression\n",
      "        - value: exposure\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: aspect qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - named thing associated with likelihood of named thing association\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      qualified_predicate:\n",
      "        name: qualified_predicate\n",
      "        description: Predicate to be used in an association when subject and object\n",
      "          qualifiers are present and the full reading of the statement requires a\n",
      "          qualification to the predicate in use in order to refine or increase the\n",
      "          specificity of the full statement reading.  This qualifier holds a relationship\n",
      "          to be used instead of that expressed by the primary predicate, in a ‘full\n",
      "          statement’ reading of the association, where qualifier-based semantics are\n",
      "          included.  This is necessary only in cases where the primary predicate does\n",
      "          not work in a full statement reading.\n",
      "        notes:\n",
      "        - 'to express the statement that “Chemical X causes increased expression of\n",
      "          Gene Y”, the core triple is read using the fields subject:ChemX, predicate:affects,\n",
      "          object:GeneY . . . and the full statement is read using the fields subject:ChemX,\n",
      "          qualified_predicate:causes, object:GeneY, object_aspect: expression, object_direction:increased.\n",
      "          The predicate ‘affects’ is needed for the core triple reading, but does\n",
      "          not make sense in the full statement reading  (because “Chemical X affects\n",
      "          increased expression of Gene Y'''' is not what we mean to say here: it causes\n",
      "          increased expression of Gene Y)'\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - predicate mapping\n",
      "        - gene regulates gene association\n",
      "        - chemical affects gene association\n",
      "        - gene affects chemical association\n",
      "        - entity to feature or disease qualifiers mixin\n",
      "        - feature or disease qualifiers to entity mixin\n",
      "      frequency_qualifier:\n",
      "        name: frequency_qualifier\n",
      "        description: a qualifier used in a phenotypic association to state how frequent\n",
      "          the phenotype is observed in the subject\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: qualifier\n",
      "        domain_of:\n",
      "        - frequency qualifier mixin\n",
      "        range: frequency value\n",
      "    defining_slots:\n",
      "    - subject\n",
      "    - object\n",
      "  Case:\n",
      "    name: Case\n",
      "    description: An individual (human) organism that has a patient role in some clinical\n",
      "      context.\n",
      "    aliases:\n",
      "    - patient\n",
      "    - proband\n",
      "    attributes:\n",
      "      in_taxon:\n",
      "        name: in_taxon\n",
      "        annotations:\n",
      "          canonical_predicate:\n",
      "            tag: canonical_predicate\n",
      "            value: true\n",
      "        description: connects an entity to its taxonomic classification. Only certain\n",
      "          kinds of entities can be taxonomically classified; see 'thing with taxon'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - instance of\n",
      "        - is organism source of gene product\n",
      "        - organism has gene\n",
      "        - gene found in organism\n",
      "        - gene product has organism source\n",
      "        exact_mappings:\n",
      "        - RO:0002162\n",
      "        - WIKIDATA_PROPERTY:P703\n",
      "        narrow_mappings:\n",
      "        - RO:0002160\n",
      "        is_a: related to at instance level\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: organism taxon\n",
      "      in_taxon_label:\n",
      "        name: in_taxon_label\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: The human readable scientific name for the taxon of the entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P225\n",
      "        is_a: node property\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: label type\n",
      "      provided_by:\n",
      "        name: provided_by\n",
      "        description: The value in this node property represents the knowledge provider\n",
      "          that created or assembled the node and all of its attributes.  Used internally\n",
      "          to represent how a particular node made its way into a knowledge provider\n",
      "          or graph.\n",
      "        is_a: node property\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        multivalued: true\n",
      "      xref:\n",
      "        name: xref\n",
      "        description: A database cross reference or alternative identifier for a NamedThing\n",
      "          or edge between two NamedThings.  This property should point to a database\n",
      "          record or webpage that supports the existence of the edge, or gives more\n",
      "          detail about the edge. This property can be used on a node or edge to provide\n",
      "          multiple URIs or CURIE cross references.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - dbxref\n",
      "        - Dbxref\n",
      "        - DbXref\n",
      "        - record_url\n",
      "        - source_record_urls\n",
      "        narrow_mappings:\n",
      "        - gff3:Dbxref\n",
      "        - gpi:DB_Xrefs\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - publication\n",
      "        - retrieval source\n",
      "        - gene\n",
      "        - gene product mixin\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      full_name:\n",
      "        name: full_name\n",
      "        description: a long-form human readable name for a thing\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        range: label type\n",
      "      synonym:\n",
      "        name: synonym\n",
      "        description: Alternate human-readable names for a thing\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - alias\n",
      "        narrow_mappings:\n",
      "        - skos:altLabel\n",
      "        - gff3:Alias\n",
      "        - AGRKB:synonyms\n",
      "        - gpi:DB_Object_Synonyms\n",
      "        - HANCESTRO:0330\n",
      "        - IAO:0000136\n",
      "        - RXNORM:has_tradename\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - gene product mixin\n",
      "        range: label type\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "  PhenotypicFeature:\n",
      "    name: PhenotypicFeature\n",
      "    id_prefixes:\n",
      "    - HP\n",
      "    - EFO\n",
      "    - NCIT\n",
      "    - UMLS\n",
      "    - MEDDRA\n",
      "    - MP\n",
      "    - ZP\n",
      "    - UPHENO\n",
      "    - APO\n",
      "    - FBcv\n",
      "    - WBPhenotype\n",
      "    - SNOMEDCT\n",
      "    - MESH\n",
      "    - XPO\n",
      "    - FYPO\n",
      "    - TO\n",
      "    description: A combination of entity and quality that makes up a phenotyping statement.\n",
      "      An observable characteristic of an individual resulting from the interaction\n",
      "      of its genotype with its molecular and physical environment.\n",
      "    examples:\n",
      "    - value: MP:0001262\n",
      "      description: decreased body weight\n",
      "    in_subset:\n",
      "    - model_organism_database\n",
      "    aliases:\n",
      "    - sign\n",
      "    - symptom\n",
      "    - phenotype\n",
      "    - trait\n",
      "    - endophenotype\n",
      "    exact_mappings:\n",
      "    - UPHENO:0001001\n",
      "    - SIO:010056\n",
      "    - WIKIDATA:Q104053\n",
      "    - UMLS:C4021819\n",
      "    - NCIT:C16977\n",
      "    - SNOMEDCT:8116006\n",
      "    - MESH:D010641\n",
      "    narrow_mappings:\n",
      "    - STY:T184\n",
      "    - WIKIDATA:Q169872\n",
      "    - WIKIDATA:Q25203551\n",
      "    - ZP:00000000\n",
      "    - FBcv:0001347\n",
      "    - HP:0000118\n",
      "    - MP:0000001\n",
      "    - WBPhenotype:0000886\n",
      "    - XPO:00000000\n",
      "    - FYPO:0000001\n",
      "    - APO:0000017\n",
      "    - TO:0000387\n",
      "    - STY:T190\n",
      "    broad_mappings:\n",
      "    - BFO:0000019\n",
      "    - PATO:0000001\n",
      "    attributes:\n",
      "      in_taxon:\n",
      "        name: in_taxon\n",
      "        annotations:\n",
      "          canonical_predicate:\n",
      "            tag: canonical_predicate\n",
      "            value: true\n",
      "        description: connects an entity to its taxonomic classification. Only certain\n",
      "          kinds of entities can be taxonomically classified; see 'thing with taxon'\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - instance of\n",
      "        - is organism source of gene product\n",
      "        - organism has gene\n",
      "        - gene found in organism\n",
      "        - gene product has organism source\n",
      "        exact_mappings:\n",
      "        - RO:0002162\n",
      "        - WIKIDATA_PROPERTY:P703\n",
      "        narrow_mappings:\n",
      "        - RO:0002160\n",
      "        is_a: related to at instance level\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: organism taxon\n",
      "      in_taxon_label:\n",
      "        name: in_taxon_label\n",
      "        annotations:\n",
      "          denormalized:\n",
      "            tag: denormalized\n",
      "            value: true\n",
      "        description: The human readable scientific name for the taxon of the entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P225\n",
      "        is_a: node property\n",
      "        domain: thing with taxon\n",
      "        domain_of:\n",
      "        - thing with taxon\n",
      "        range: label type\n",
      "      provided_by:\n",
      "        name: provided_by\n",
      "        description: The value in this node property represents the knowledge provider\n",
      "          that created or assembled the node and all of its attributes.  Used internally\n",
      "          to represent how a particular node made its way into a knowledge provider\n",
      "          or graph.\n",
      "        is_a: node property\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        multivalued: true\n",
      "      xref:\n",
      "        name: xref\n",
      "        description: A database cross reference or alternative identifier for a NamedThing\n",
      "          or edge between two NamedThings.  This property should point to a database\n",
      "          record or webpage that supports the existence of the edge, or gives more\n",
      "          detail about the edge. This property can be used on a node or edge to provide\n",
      "          multiple URIs or CURIE cross references.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - dbxref\n",
      "        - Dbxref\n",
      "        - DbXref\n",
      "        - record_url\n",
      "        - source_record_urls\n",
      "        narrow_mappings:\n",
      "        - gff3:Dbxref\n",
      "        - gpi:DB_Xrefs\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - publication\n",
      "        - retrieval source\n",
      "        - gene\n",
      "        - gene product mixin\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      full_name:\n",
      "        name: full_name\n",
      "        description: a long-form human readable name for a thing\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        range: label type\n",
      "      synonym:\n",
      "        name: synonym\n",
      "        description: Alternate human-readable names for a thing\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - alias\n",
      "        narrow_mappings:\n",
      "        - skos:altLabel\n",
      "        - gff3:Alias\n",
      "        - AGRKB:synonyms\n",
      "        - gpi:DB_Object_Synonyms\n",
      "        - HANCESTRO:0330\n",
      "        - IAO:0000136\n",
      "        - RXNORM:has_tradename\n",
      "        is_a: node property\n",
      "        domain: named thing\n",
      "        domain_of:\n",
      "        - named thing\n",
      "        - gene product mixin\n",
      "        range: label type\n",
      "        multivalued: true\n",
      "      id:\n",
      "        name: id\n",
      "        description: A unique identifier for an entity. Must be either a CURIE shorthand\n",
      "          for a URI or a complete URI\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        exact_mappings:\n",
      "        - AGRKB:primaryId\n",
      "        - gff3:ID\n",
      "        - gpi:DB_Object_ID\n",
      "        domain: entity\n",
      "        identifier: true\n",
      "        domain_of:\n",
      "        - ontology class\n",
      "        - entity\n",
      "        required: true\n",
      "      iri:\n",
      "        name: iri\n",
      "        description: An IRI for an entity. This is determined by the id using expansion\n",
      "          rules.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - WIKIDATA_PROPERTY:P854\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        range: iri type\n",
      "      category:\n",
      "        name: category\n",
      "        description: Name of the high level ontology class in which this entity is\n",
      "          categorized. Corresponds to the label for the biolink entity type class.\n",
      "          In a neo4j database this MAY correspond to the neo4j label tag. In an RDF\n",
      "          database it should be a biolink model class URI. This field is multi-valued.\n",
      "          It should include values for ancestors of the biolink class; for example,\n",
      "          a protein such as Shh would have category values `biolink:Protein`, `biolink:GeneProduct`,\n",
      "          `biolink:MolecularEntity`. In an RDF database, nodes will typically have\n",
      "          an rdf:type triples. This can be to the most specific biolink class, or\n",
      "          potentially to a class more specific than something in biolink. For example,\n",
      "          a sequence feature `f` may have a rdf:type assertion to a SO class such\n",
      "          as TF_binding_site, which is more specific than anything in biolink. Here\n",
      "          we would have categories {biolink:GenomicEntity, biolink:MolecularEntity,\n",
      "          biolink:NamedThing}\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        is_a: type\n",
      "        domain: entity\n",
      "        designates_type: true\n",
      "        is_class_field: true\n",
      "        range: uriorcurie\n",
      "        multivalued: true\n",
      "      type:\n",
      "        name: type\n",
      "        exact_mappings:\n",
      "        - gff3:type\n",
      "        - gpi:DB_Object_Type\n",
      "        domain: entity\n",
      "        slot_uri: rdf:type\n",
      "        domain_of:\n",
      "        - entity\n",
      "        multivalued: true\n",
      "      name:\n",
      "        name: name\n",
      "        description: A human-readable name for an attribute or entity.\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        - samples\n",
      "        aliases:\n",
      "        - label\n",
      "        - display name\n",
      "        - title\n",
      "        exact_mappings:\n",
      "        - gff3:Name\n",
      "        - gpi:DB_Object_Name\n",
      "        narrow_mappings:\n",
      "        - dct:title\n",
      "        - WIKIDATA_PROPERTY:P1476\n",
      "        domain: entity\n",
      "        slot_uri: rdfs:label\n",
      "        domain_of:\n",
      "        - attribute\n",
      "        - entity\n",
      "        - macromolecular machine mixin\n",
      "        range: label type\n",
      "      description:\n",
      "        name: description\n",
      "        description: a human-readable description of an entity\n",
      "        in_subset:\n",
      "        - translator_minimal\n",
      "        aliases:\n",
      "        - definition\n",
      "        exact_mappings:\n",
      "        - IAO:0000115\n",
      "        - skos:definitions\n",
      "        narrow_mappings:\n",
      "        - gff3:Description\n",
      "        slot_uri: dct:description\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: narrative text\n",
      "      has_attribute:\n",
      "        name: has_attribute\n",
      "        description: connects any entity to an attribute\n",
      "        in_subset:\n",
      "        - samples\n",
      "        exact_mappings:\n",
      "        - SIO:000008\n",
      "        close_mappings:\n",
      "        - OBI:0001927\n",
      "        narrow_mappings:\n",
      "        - OBAN:association_has_subject_property\n",
      "        - OBAN:association_has_object_property\n",
      "        - CPT:has_possibly_included_panel_element\n",
      "        - DRUGBANK:category\n",
      "        - EFO:is_executed_in\n",
      "        - HANCESTRO:0301\n",
      "        - LOINC:has_action_guidance\n",
      "        - LOINC:has_adjustment\n",
      "        - LOINC:has_aggregation_view\n",
      "        - LOINC:has_approach_guidance\n",
      "        - LOINC:has_divisor\n",
      "        - LOINC:has_exam\n",
      "        - LOINC:has_method\n",
      "        - LOINC:has_modality_subtype\n",
      "        - LOINC:has_object_guidance\n",
      "        - LOINC:has_scale\n",
      "        - LOINC:has_suffix\n",
      "        - LOINC:has_time_aspect\n",
      "        - LOINC:has_time_modifier\n",
      "        - LOINC:has_timing_of\n",
      "        - NCIT:R88\n",
      "        - NCIT:eo_disease_has_property_or_attribute\n",
      "        - NCIT:has_data_element\n",
      "        - NCIT:has_pharmaceutical_administration_method\n",
      "        - NCIT:has_pharmaceutical_basic_dose_form\n",
      "        - NCIT:has_pharmaceutical_intended_site\n",
      "        - NCIT:has_pharmaceutical_release_characteristics\n",
      "        - NCIT:has_pharmaceutical_state_of_matter\n",
      "        - NCIT:has_pharmaceutical_transformation\n",
      "        - NCIT:is_qualified_by\n",
      "        - NCIT:qualifier_applies_to\n",
      "        - NCIT:role_has_domain\n",
      "        - NCIT:role_has_range\n",
      "        - INO:0000154\n",
      "        - HANCESTRO:0308\n",
      "        - OMIM:has_inheritance_type\n",
      "        - orphanet:C016\n",
      "        - orphanet:C017\n",
      "        - RO:0000053\n",
      "        - RO:0000086\n",
      "        - RO:0000087\n",
      "        - SNOMED:has_access\n",
      "        - SNOMED:has_clinical_course\n",
      "        - SNOMED:has_count_of_base_of_active_ingredient\n",
      "        - SNOMED:has_dose_form_administration_method\n",
      "        - SNOMED:has_dose_form_release_characteristic\n",
      "        - SNOMED:has_dose_form_transformation\n",
      "        - SNOMED:has_finding_context\n",
      "        - SNOMED:has_finding_informer\n",
      "        - SNOMED:has_inherent_attribute\n",
      "        - SNOMED:has_intent\n",
      "        - SNOMED:has_interpretation\n",
      "        - SNOMED:has_laterality\n",
      "        - SNOMED:has_measurement_method\n",
      "        - SNOMED:has_method\n",
      "        - SNOMED:has_priority\n",
      "        - SNOMED:has_procedure_context\n",
      "        - SNOMED:has_process_duration\n",
      "        - SNOMED:has_property\n",
      "        - SNOMED:has_revision_status\n",
      "        - SNOMED:has_scale_type\n",
      "        - SNOMED:has_severity\n",
      "        - SNOMED:has_specimen\n",
      "        - SNOMED:has_state_of_matter\n",
      "        - SNOMED:has_subject_relationship_context\n",
      "        - SNOMED:has_surgical_approach\n",
      "        - SNOMED:has_technique\n",
      "        - SNOMED:has_temporal_context\n",
      "        - SNOMED:has_time_aspect\n",
      "        - SNOMED:has_units\n",
      "        - UMLS:has_structural_class\n",
      "        - UMLS:has_supported_concept_property\n",
      "        - UMLS:has_supported_concept_relationship\n",
      "        - UMLS:may_be_qualified_by\n",
      "        domain: entity\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: attribute\n",
      "        multivalued: true\n",
      "      deprecated:\n",
      "        name: deprecated\n",
      "        description: A boolean flag indicating that an entity is no longer considered\n",
      "          current or valid.\n",
      "        exact_mappings:\n",
      "        - oboInOwl:ObsoleteClass\n",
      "        domain_of:\n",
      "        - entity\n",
      "        range: boolean\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# print the content of the new schema in LinkML YAML format to view here in the notebook\n",
    "yaml_content = yaml_dumper.dumps(target_schema_obj)  # Serialize to a string\n",
    "print(yaml_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d65b0bfa68ba6d",
   "metadata": {},
   "source": [
    "Notes:\n",
    "* notice that if we remove a parent class from this, e.g. remove \"NamedThing\" - the `is_a` path in the descendent classes will be absent.  this is to prevent unreachable element errors.\n",
    "* notice that helpful defaults are brought in like prefixes, descriptions, aliases, mappings, etc.\n",
    "  * there will likely be cases where metamodel elements in LinkML are not automatically transferred to the derived schema (as will all our generators, we are working towards feature parity).  \n",
    "* notice that the transformation automatically makes what were, independent `slot` definitions in Biolink Model into `attributes`.  These are more or less functionally equivalent in LinkML, however if you want to specify a slot that can be reused outside of a particular class, it is still best practice to do so with a slot definition rather than an attribute as to not repeat slot definitions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9addb6853dbd3759",
   "metadata": {},
   "source": [
    "### Use derived schema to generate different serializations of the derived model\n",
    "\n",
    "Now that we have a YAML dump of our derived model, we can use standard LinkML tooling to produce:\n",
    "* python dataclasses and pydantic models of our derived schema\n",
    "* navigate our derived schema with a SchemaView instance\n",
    "* create and deploy automated documentation with the derived schema (see https://github.com/linkml/linkml-project-cookiecutter for more details on using the derived schema in a standard setup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f8ccc4af61d75af7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception: range: biological sequence\n"
     ]
    }
   ],
   "source": [
    "!gen-pydantic biolink-subset.yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c811e7ae6de6e3a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
